// @pepperi-addons/ngx-lib/esm2015/form/indicators.component.js
import { Component, Input, ChangeDetectionStrategy } from '@angular/core';
export class PepIndicatorsComponent {
    constructor() {
        this.controlType = 'indicators';
        this.layoutType = 'table';
    }
}
PepIndicatorsComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-indicators',
                template: "<div class=\"indicators-container\" [ngClass]=\"{ 'is-inside-table': layoutType === 'table' }\">\n    <ng-container *ngFor=\"let value of value?.split(';')\">\n        <pep-icon name=\"indicator_dot_placeholder\" class=\"pull-left flip\" [fill]=\"value\"></pep-icon>\n    </ng-container>\n    <ng-container *ngIf=\"key != 'ItemIndicatorsWithoutCampaign' && value?.split(';').length < 4\">\n        <pep-icon name=\"indicator_dot_placeholder\" class=\"pull-left flip\"></pep-icon>\n    </ng-container>\n</div>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{height:inherit;display:grid}:host>*{align-self:center}.indicators-container{height:2.5rem}.indicators-container pep-icon{height:1rem;width:1rem;margin:.05rem}.indicators-container.is-inside-table{height:1rem;display:flex;justify-content:center;align-items:center}"]
            },] }
];
PepIndicatorsComponent.propDecorators = {
    key: [{ type: Input }],
    value: [{ type: Input }],
    layoutType: [{ type: Input }]
};
