// @pepperi-addons/ngx-lib/esm2015/core/layout/rtl.directive.js
import { Directive, Renderer2, ElementRef } from '@angular/core';
import { PepLayoutService } from './layout.service';
export class PepRtlClassDirective {
    constructor(layoutService, renderer, element) {
        this.layoutService = layoutService;
        this.renderer = renderer;
        this.element = element;
    }
    ngOnInit() {
        const isRtl = this.layoutService.isRtl();
        if (isRtl) {
            this.renderer.addClass(this.element.nativeElement, 'rtl');
        }
    }
}
PepRtlClassDirective.decorators = [
    { type: Directive, args: [{
                selector: '[pepRtlClass]',
            },] }
];
PepRtlClassDirective.ctorParameters = () => [
    { type: PepLayoutService },
    { type: Renderer2 },
    { type: ElementRef }
];
export class PepRtlDirectionDirective {
    constructor(layoutService, renderer, element) {
        this.layoutService = layoutService;
        this.renderer = renderer;
        this.element = element;
    }
    ngOnInit() {
        const isRtl = this.layoutService.isRtl();
        this.renderer.setAttribute(this.element.nativeElement, 'dir', isRtl ? 'rtl' : 'ltr');
    }
}
PepRtlDirectionDirective.decorators = [
    { type: Directive, args: [{
                selector: '[pepRtlDirection]',
            },] }
];
PepRtlDirectionDirective.ctorParameters = () => [
    { type: PepLayoutService },
    { type: Renderer2 },
    { type: ElementRef }
];
