// @pepperi-addons/ngx-lib/esm2015/core/common/services/window-scrolling.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class PepWindowScrollingService {
    constructor() {
        this.styleTag = this.buildStyleElement();
    }
    // Disable the scrolling feature on the main viewport.
    disable() {
        document.body.appendChild(this.styleTag);
    }
    // Re-enable the scrolling feature on the main viewport.
    enable() {
        document.body.removeChild(this.styleTag);
    }
    // Return a Style element that will prevent scrolling on the body.
    buildStyleElement() {
        const style = document.createElement('style');
        style.type = 'text/css';
        style.setAttribute('data-debug', 'Injected by WindowScrolling service.');
        style.textContent = `
            body {
                overflow: hidden !important ;
            }
        `;
        return style;
    }
}
PepWindowScrollingService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PepWindowScrollingService_Factory() { return new PepWindowScrollingService(); }, token: PepWindowScrollingService, providedIn: "root" });
PepWindowScrollingService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
PepWindowScrollingService.ctorParameters = () => [];
