// @pepperi-addons/ngx-lib/esm2015/core/common/directives/auto-width.directive.js
import { Directive, ElementRef, Input, Renderer2, } from '@angular/core';
export class PepInputAutoWidthDirective {
    constructor(renderer, el) {
        this.renderer = renderer;
        this.el = el;
        this.includePadding = true;
        this.input = null;
    }
    ngAfterViewInit() {
        if (this.el.nativeElement.tagName.toLowerCase() === 'input') {
            this.input = this.el;
        }
        else {
            const list = this.el.nativeElement.querySelectorAll('input');
            this.input = list.length > 0 ? list[0] : null;
        }
        if (this.input) {
            this.input.addEventListener('blur', this.setWidthUsingText.bind(this));
            this.input.addEventListener('keyup', this.setWidthUsingText.bind(this));
        }
        this.setWidthUsingText();
    }
    ngOnDestroy() {
        this.input.removeEventListener('blur', this.setWidthUsingText.bind(this));
        this.input.removeEventListener('keyup', this.setWidthUsingText.bind(this));
    }
    get paddingWidth() {
        return this.includePadding
            ? this._sumPropertyValues(['padding-left', 'padding-right'])
            : 0;
    }
    get style() {
        return getComputedStyle(this.input, '');
    }
    setWidth(width) {
        this.renderer.setStyle(this.input, 'width', width + 'px');
    }
    setWidthUsingText() {
        if (this.input) {
            const text = this.input.value;
            this.setWidth(this.textWidth(text) + this.paddingWidth);
        }
    }
    _sumPropertyValues(properties) {
        return properties
            .map((property) => parseInt(this.style.getPropertyValue(property), 10))
            .reduce((a, b) => a + b, 0);
    }
    textWidth(value) {
        const ctx = this.renderer.createElement('canvas').getContext('2d');
        const { fontStyle, fontVariant, fontWeight, fontSize, fontFamily, } = this.style;
        // font string format: {normal, normal, 700, 20px, Roboto, "Helvetica Neue", sans-serif}
        ctx.font =
            fontStyle +
                ' ' +
                fontVariant +
                ' ' +
                fontWeight +
                ' ' +
                fontSize +
                ' ' +
                fontFamily;
        return ctx ? ctx.measureText(value).width : 0;
    }
}
PepInputAutoWidthDirective.decorators = [
    { type: Directive, args: [{
                selector: '[pepAutoWidth]',
            },] }
];
PepInputAutoWidthDirective.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
PepInputAutoWidthDirective.propDecorators = {
    includePadding: [{ type: Input }]
};
