// @pepperi-addons/ngx-lib/fesm2015/pepperi-addons-ngx-lib-field-title.js
import { Component, Input, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatCommonModule } from '@angular/material/core';
import { MatIconModule } from '@angular/material/icon';
import { MatFormFieldModule } from '@angular/material/form-field';
import { pepIconSystemMust, PepIconModule, PepIconRegistry } from '@pepperi-addons/ngx-lib/icon';
import { DEFAULT_HORIZONTAL_ALIGNMENT } from '@pepperi-addons/ngx-lib';

class PepFieldTitleComponent {
    constructor() {
        this.label = '';
        this.mandatory = false;
        this.disabled = false;
        this.maxFieldCharacters = 0;
        this.hint = '';
        this.xAlignment = DEFAULT_HORIZONTAL_ALIGNMENT;
        this.showTitle = true;
        this.inputLength = 0;
    }
}
PepFieldTitleComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-field-title',
                template: "<div class=\"mat-form-title\" [ngClass]=\"{disable: disabled}\" [ngStyle]=\"{'text-align': xAlignment }\">\n    <ng-container *ngIf=\"xAlignment == 'right' then rightTemplate else leftTemplate\">\n    </ng-container>\n    <ng-template #rightTemplate>\n        <ng-container *ngTemplateOutlet=\"hintTemplate\"></ng-container>\n        <ng-container *ngTemplateOutlet=\"titleTemplate\"></ng-container>\n        <ng-container *ngTemplateOutlet=\"requiredTemplate\"></ng-container>\n    </ng-template>\n    <ng-template #leftTemplate>\n        <ng-container *ngTemplateOutlet=\"requiredTemplate\"></ng-container>\n        <ng-container *ngTemplateOutlet=\"titleTemplate\"></ng-container>\n        <ng-container *ngTemplateOutlet=\"hintTemplate\"></ng-container>\n    </ng-template>\n\n    <ng-template #requiredTemplate>\n        <mat-icon *ngIf=\"mandatory\" class=\"pep-spacing-element\">\n            <pep-icon name=\"system_must\"></pep-icon>\n        </mat-icon>\n        <!-- For the grid template () -->\n        <span *ngIf=\"!mandatory\"></span>\n    </ng-template>\n    <ng-template #titleTemplate>\n        <mat-label *ngIf=\"showTitle\" [title]=\"label\" class=\"body-xs\">\n            {{ label }}\n        </mat-label>\n        <!-- For the grid template () -->\n        <span *ngIf=\"!showTitle\"></span>\n    </ng-template>\n    <ng-template #hintTemplate>\n        <ng-container *ngIf=\"hint?.length > 0 then hintTextTemplate else maxCharactersTemplate\">\n        </ng-container>\n        <ng-template #hintTextTemplate>\n            <span class=\"body-xs\">\n                &nbsp;{{ hint }}\n            </span>\n        </ng-template>\n        <ng-template #maxCharactersTemplate>\n            <span *ngIf=\"maxFieldCharacters > 0\" class=\"body-xs\">\n                &nbsp;{{ inputLength || 0 }}/{{ maxFieldCharacters }}\n            </span>\n            <!-- For the grid template () -->\n            <span *ngIf=\"maxFieldCharacters === 0\"></span>\n        </ng-template>\n    </ng-template>\n</div>",
                styles: [":host{display:grid;width:inherit}"]
            },] }
];
PepFieldTitleComponent.propDecorators = {
    label: [{ type: Input }],
    mandatory: [{ type: Input }],
    disabled: [{ type: Input }],
    maxFieldCharacters: [{ type: Input }],
    hint: [{ type: Input }],
    xAlignment: [{ type: Input }],
    showTitle: [{ type: Input }],
    inputLength: [{ type: Input }]
};

class PepFieldTitleModule {
    constructor(pepIconRegistry) {
        this.pepIconRegistry = pepIconRegistry;
        this.pepIconRegistry.registerIcons([pepIconSystemMust]);
    }
}
PepFieldTitleModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    // Material modules,
                    MatCommonModule,
                    MatIconModule,
                    MatFormFieldModule,
                    // ngx-lib modules
                    PepIconModule,
                ],
                exports: [PepFieldTitleComponent],
                declarations: [PepFieldTitleComponent],
            },] }
];
PepFieldTitleModule.ctorParameters = () => [
    { type: PepIconRegistry }
];

/*
 * Public API Surface of ngx-lib/field-title
 */

/**
 * Generated bundle index. Do not edit.
 */

export { PepFieldTitleComponent, PepFieldTitleModule };
