// @pepperi-addons/ngx-lib/fesm2015/pepperi-addons-ngx-lib-separator.js
import { Component, ChangeDetectionStrategy, Renderer2, ElementRef, Input, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DEFAULT_HORIZONTAL_ALIGNMENT, PepCustomizationService } from '@pepperi-addons/ngx-lib';

class PepSeparatorComponent {
    constructor(renderer, element) {
        this.renderer = renderer;
        this.element = element;
        this.key = '';
        this.label = '';
        this.xAlignment = DEFAULT_HORIZONTAL_ALIGNMENT;
        this.controlType = 'separator';
        this.form = null;
        this.layoutType = 'form';
        this._visible = true;
        this.standAlone = false;
    }
    set visible(visible) {
        this._visible = visible;
        if (visible) {
            this.renderer.removeClass(this.element.nativeElement, 'hidden-element');
        }
        else {
            this.renderer.addClass(this.element.nativeElement, 'hidden-element');
        }
    }
    get visible() {
        return this._visible;
    }
    ngOnInit() {
        if (this.form === null) {
            this.standAlone = true;
            this.renderer.addClass(this.element.nativeElement, PepCustomizationService.STAND_ALONE_FIELD_CLASS_NAME);
        }
    }
}
PepSeparatorComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-separator',
                template: "<ng-container *ngIf=\"layoutType === 'form'\">\n    <div class=\"pep-separator-field wrap pep-border-bottom align-{{ xAlignment }}\" title=\"{{ label }}\">\n        <span class=\"block-with-text body-md\">{{ label }}</span>\n    </div>\n</ng-container>\n\n<ng-container *ngIf=\"layoutType === 'card'\">\n    <div class=\"pep-separator-field pep-small-separator-field pep-border-bottom text-align-{{ xAlignment }}\"\n        dir=\"{{ xAlignment == 'right' ? 'rtl' : 'ltr' }}\" title=\"{{ label }}\">\n        <span class=\"small-block-with-text body-sm\">{{ label }}</span>\n    </div>\n</ng-container>\n\n<ng-container *ngIf=\"layoutType === 'table'\">\n    <label class=\"cl\" title=\"{{ label }}\">{{ label }}</label>\n</ng-container>",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [":host{height:inherit;display:grid}:host>*{align-self:center}"]
            },] }
];
PepSeparatorComponent.ctorParameters = () => [
    { type: Renderer2 },
    { type: ElementRef }
];
PepSeparatorComponent.propDecorators = {
    key: [{ type: Input }],
    label: [{ type: Input }],
    xAlignment: [{ type: Input }],
    form: [{ type: Input }],
    layoutType: [{ type: Input }],
    visible: [{ type: Input }]
};

class PepSeparatorModule {
}
PepSeparatorModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                exports: [PepSeparatorComponent],
                declarations: [PepSeparatorComponent],
            },] }
];

/*
 * Public API Surface of ngx-lib/separator
 */

/**
 * Generated bundle index. Do not edit.
 */

export { PepSeparatorComponent, PepSeparatorModule };
