// @pepperi-addons/ngx-lib/esm2015/smart-filters/common/filter-actions.component.js
import { Component, EventEmitter, Input, Output } from '@angular/core';
export class PepFilterActionsComponent {
    constructor() {
        this.clearClick = new EventEmitter();
        this.applyClick = new EventEmitter();
    }
    clear() {
        this.clearClick.emit();
    }
    apply() {
        this.applyClick.emit();
    }
}
PepFilterActionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-filter-actions',
                template: "<div [formGroup]=\"form\" class=\"filter-actions pep-spacing-element-negative\">\n    <button mat-button class=\"pep-spacing-element pep-button sm weak\" (click)=\"clear()\">\n        {{ 'ACTIONS.CLEAR' | translate }}\n    </button>\n    <button mat-button class=\"pep-spacing-element pep-button sm strong\" (click)=\"apply()\" [disabled]=\"form.invalid\">\n        {{ 'ACTIONS.APPLY' | translate }}\n    </button>\n</div>",
                styles: [".filter-actions{margin:1rem 0 calc(1rem / 2);margin:var(--pep-spacing-lg,1rem) 0 calc(var(--pep-spacing-lg, 1rem) / 2);display:flex;justify-content:flex-end}"]
            },] }
];
PepFilterActionsComponent.propDecorators = {
    form: [{ type: Input }],
    clearClick: [{ type: Output }],
    applyClick: [{ type: Output }]
};
