// @pepperi-addons/ngx-lib/esm2015/smart-filters/number-filter/number-filter.component.js
import { Component } from '@angular/core';
import { BaseFilterComponent } from '../common/model/base-filter-component';
import { PepSmartFilterOperators, } from '../common/model/operator';
import { Validators } from '@angular/forms';
import { distinctUntilChanged } from 'rxjs/operators';
export class PepNumberFilterComponent extends BaseFilterComponent {
    constructor() {
        super(...arguments);
        this.PepSmartFilterOperators = PepSmartFilterOperators;
        this.chooseTypeOptions = [];
    }
    ngOnInit() {
        this.firstControl.valueChanges
            .pipe(this.getDestroyer(), distinctUntilChanged())
            .subscribe(() => {
            this.setFieldsStateAndValidators();
        });
        this.secondControl.valueChanges
            .pipe(this.getDestroyer(), distinctUntilChanged())
            .subscribe(() => {
            this.setFieldsStateAndValidators();
        });
    }
    // Override
    getDefaultOperator() {
        return PepSmartFilterOperators.Equals;
    }
    // Override
    getFilterValue() {
        const filterValue = {
            first: this.firstControl.value,
        };
        if (this.operator === PepSmartFilterOperators.NumberRange) {
            filterValue['second'] = this.secondControl.value;
        }
        return filterValue;
    }
    // Override
    loadOperatorsOptions() {
        this.chooseTypeOptions = this.operators.map((operator) => {
            return {
                key: operator.id,
                value: this.translate.instant(`${this.OPERATORS_TRANSLATION_PREFIX}.${operator.name}`),
            };
        });
    }
    // Override
    setFieldsStateAndValidators() {
        if (this.operator === PepSmartFilterOperators.NumberRange) {
            const firstValidators = [Validators.required];
            if (this.secondControl.value) {
                firstValidators.push(this.validator.isLessThan(this.secondControl));
            }
            this.firstControl.setValidators(firstValidators);
            this.secondControl.enable();
            const secondValidators = [Validators.required];
            if (this.firstControl.value) {
                secondValidators.push(this.validator.isGreaterThan(this.firstControl));
            }
            this.secondControl.setValidators(secondValidators);
        }
        else {
            super.setFieldsStateAndValidators();
        }
    }
    onOperatorChanged(value) {
        const operator = Object.values(PepSmartFilterOperators).find((operator) => operator.id === value);
        this.operator = operator;
    }
}
PepNumberFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-number-filter',
                template: "<div [formGroup]=\"form\">\n    <ng-container>\n        <pep-select [key]=\"'typeOperator'\" [value]=\"operator.id\" [label]=\"'SMART_FILTERS.TYPE' | translate\"\n            [emptyOption]=\"false\" [options]=\"chooseTypeOptions\" (valueChange)=\"onOperatorChanged($event)\">\n        </pep-select>\n    </ng-container>\n    <div *ngIf=\"operator !== PepSmartFilterOperators.NumberRange\">\n        <!-- Amount -->\n        <pep-textbox [form]=\"form\" [key]=\"firstControlKey\" [label]=\"'SMART_FILTERS.AMOUNT' | translate\"\n            [renderError]=\"false\" [type]=\"field.type\" [formattedValue]=\"firstControl.value\"\n            [value]=\"firstControl.value\">\n        </pep-textbox>\n    </div>\n    <div *ngIf=\"operator === PepSmartFilterOperators.NumberRange\" fxLayout=\"row\" fxLayout.xs=\"column\"\n        fxLayoutGap=\"1rem\">\n        <!-- Min number -->\n        <pep-textbox [form]=\"form\" [key]=\"firstControlKey\" [label]=\"'SMART_FILTERS.MIN' | translate\"\n            [renderError]=\"false\" fxFlex=\"50%\" [type]=\"field.type\" [formattedValue]=\"firstControl.value\"\n            [value]=\"firstControl.value\" [maxValue]=\"secondControl.value\">\n        </pep-textbox>\n        <!-- Max number -->\n        <pep-textbox [form]=\"form\" [key]=\"secondControlKey\" [label]=\"'SMART_FILTERS.MAX' | translate\"\n            [renderError]=\"false\" fxFlex=\"50%\" [type]=\"field.type\" [formattedValue]=\"secondControl.value\"\n            [value]=\"secondControl.value\" [minValue]=\"firstControl.value\">\n        </pep-textbox>\n    </div>\n</div>",
                styles: [""]
            },] }
];
