// @pepperi-addons/ngx-lib/esm2015/core/customization/customization.service.js
import { Injectable } from '@angular/core';
import { FormBuilder } from '@angular/forms';
import { BehaviorSubject } from 'rxjs';
import { PepSessionService } from '../common/services/session.service';
import { PepFileService } from '../common/services/file.service';
import * as i0 from "@angular/core";
import * as i1 from "../common/services/session.service";
import * as i2 from "../common/services/file.service";
import * as i3 from "@angular/forms";
export class PepCustomizationService {
    constructor(sessionService, fileService, fb) {
        this.sessionService = sessionService;
        this.fileService = fileService;
        this.fb = fb;
        this.hasCustomHomepage = null;
        this.hasCustomHeader = null;
        this.customHeaderHeight = PepCustomizationService.DEFAULT_HEADER_HEIGHT;
        this.mainContHeight = 0;
        this.footerHeight = new BehaviorSubject(0);
        this.settingsWidth = new BehaviorSubject(0);
        this.hideFooter();
        //
    }
    getDefaultFromGroup(field, withValidators = true) {
        const validators = withValidators ? field.getValidators() : [];
        const group = {};
        group[field.key] = [
            { value: field.value, disabled: field.disabled },
            validators,
        ];
        return this.fb.group(group);
    }
    getFormControl(form, fieldKey, parentFieldKey = null) {
        let formControl = null;
        if (form && form.controls) {
            if (parentFieldKey === null) {
                formControl = form.controls[fieldKey];
            }
            else {
                formControl = form.controls[parentFieldKey].get(fieldKey);
            }
        }
        return formControl;
    }
    updateFormFieldValue(form, fieldKey, value = '', parentFieldKey = null) {
        const formControl = this.getFormControl(form, fieldKey, parentFieldKey);
        if (formControl) {
            formControl.setValue(value);
        }
    }
    updateFormField(form, field, value = '', parentField = null) {
        const formControl = this.getFormControl(form, field.key, parentField === null || parentField === void 0 ? void 0 : parentField.key);
        if (formControl) {
            formControl.setValue(value);
            // If disabled has changed.
            if (formControl.disabled !== field.disabled) {
                field.disabled ? formControl.disable() : formControl.enable();
                const validators = field.getValidators();
                formControl.setValidators(validators);
            }
        }
    }
    calculateFormFieldHeight(withTitle = true, rowSpan = 1, standAlone = false) {
        const themeVars = this.getThemeVariables();
        const rowFieldHeight = this.getNumberThemeVariable(themeVars, PepCustomizationService.FORM_FIELD_HEIGHT_KEY);
        const rowFieldTitleHeight = this.getNumberThemeVariable(themeVars, PepCustomizationService.FORM_FIELD_TITLE_HEIGHT_KEY);
        const rowFieldSpacingBottom = this.getNumberThemeVariable(themeVars, PepCustomizationService.FORM_SPACEING_KEY);
        let fieldHeight = 0;
        if (rowSpan === 1) {
            fieldHeight =
                (withTitle ? rowFieldTitleHeight : 0) + rowFieldHeight;
        }
        else {
            const rowSpanCalc = withTitle ? rowSpan : rowSpan - 1;
            fieldHeight =
                rowSpan * rowFieldHeight +
                    rowSpanCalc * rowFieldTitleHeight +
                    (standAlone ? 0 : rowSpanCalc * rowFieldSpacingBottom);
        }
        return fieldHeight;
    }
    calculateCardRowsHeight(rowsCount = 1, withSpacing = true) {
        const themeVars = this.getThemeVariables();
        const cardFieldHeight = this.getNumberThemeVariable(themeVars, PepCustomizationService.CARD_FIELD_HEIGHT_KEY);
        const cardFieldSpacing = this.getNumberThemeVariable(themeVars, PepCustomizationService.CARD_SPACEING_KEY);
        const fieldHeight = rowsCount * cardFieldHeight +
            (rowsCount - 1) * (withSpacing ? cardFieldSpacing : 0);
        return fieldHeight;
    }
    calculateTableRowsHeight(rowsCount = 1, withSpacing = true) {
        const themeVars = this.getThemeVariables();
        const tableFieldHeight = this.getNumberThemeVariable(themeVars, PepCustomizationService.TABLE_FIELD_HEIGHT_KEY);
        const tableFieldSpacing = this.getNumberThemeVariable(themeVars, PepCustomizationService.TABLE_SPACEING_KEY);
        const fieldHeight = rowsCount *
            (tableFieldHeight + (withSpacing ? tableFieldSpacing * 2 : 0));
        return fieldHeight;
    }
    calculateFieldHeight(layoutType = 'form', rowSpan, standAlone) {
        let fieldHeight = 'inherit'; // Default for card (with no title)
        if (layoutType === 'table') {
            fieldHeight = this.getThemeVariable(PepCustomizationService.TABLE_FIELD_HEIGHT_KEY);
        }
        else if (layoutType === 'card') {
            fieldHeight =
                this.calculateCardRowsHeight(rowSpan, !standAlone) +
                    PepCustomizationService.REM_STRING;
        }
        else {
            // Form
            fieldHeight =
                this.calculateFormFieldHeight(false, rowSpan, standAlone) +
                    PepCustomizationService.REM_STRING;
        }
        return fieldHeight;
    }
    getBrandingTheme() {
        const themeObj = this.sessionService.getObject(PepCustomizationService.USER_THEME);
        return (themeObj && themeObj.Theme) || 'default-theme'; // 'light-theme';
    }
    getDefaultThemeLayoutVariables(themeVars) {
        // Declare default screen sizes.
        themeVars['--pep-screen-max-size-2xs'] =
            '460' + PepCustomizationService.PX_STRING;
        themeVars['--pep-screen-max-size-xs'] =
            '767' + PepCustomizationService.PX_STRING;
        themeVars['--pep-screen-max-size-sm'] =
            '991' + PepCustomizationService.PX_STRING;
        themeVars['--pep-screen-max-size-md'] =
            '1199' + PepCustomizationService.PX_STRING;
        themeVars['--pep-screen-max-size-lg'] =
            '1366' + PepCustomizationService.PX_STRING;
        themeVars['--pep-screen-max-size-xl'] =
            '1920' + PepCustomizationService.PX_STRING;
        // Declare default top bar.
        themeVars[PepCustomizationService.SIDE_BAR_WIDTH_KEY] =
            '16' + PepCustomizationService.REM_STRING;
        // Declare default top bar.
        themeVars[PepCustomizationService.TOP_BAR_SPACING_TOP_KEY] =
            '1.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.TOP_BAR_SPACING_BOTTOM_KEY] =
            '0.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.TOP_BAR_FIELD_HEIGHT_KEY] =
            '2.5' + PepCustomizationService.REM_STRING;
        // Declare default footer.
        themeVars[PepCustomizationService.FOOTER_BAR_SPACING_TOP_KEY] =
            '0.75' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FOOTER_BAR_SPACING_BOTTOM_KEY] =
            '1.25' + PepCustomizationService.REM_STRING;
    }
    getDefaultThemeBorderRadiusVariables(themeVars) {
        themeVars[PepCustomizationService.BORDER_RADIUS_SM_KEY] =
            '0.125' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.BORDER_RADIUS_MD_KEY] =
            '0.25' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.BORDER_RADIUS_LG_KEY] =
            '0.5' + PepCustomizationService.REM_STRING;
    }
    getDefaultThemeSpacingVariables(themeVars) {
        themeVars[PepCustomizationService.SPACING_SIZE_2XS_KEY] =
            '0.125' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_XS_KEY] =
            '0.25' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_SM_KEY] =
            '0.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_MD_KEY] =
            '0.75' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_LG_KEY] =
            '1' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_XL_KEY] =
            '1.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_2XL_KEY] =
            '2' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_3XL_KEY] =
            '2.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_4XL_KEY] =
            '3' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_5XL_KEY] =
            '3.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_6XL_KEY] =
            '5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_7XL_KEY] =
            '6' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_8XL_KEY] =
            '8' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.SPACING_SIZE_9XL_KEY] =
            '16' + PepCustomizationService.REM_STRING;
    }
    getDefaultThemeFontsVariables(themeVars) {
        // Title font family
        themeVars[PepCustomizationService.FONT_FAMILY_TITLE_KEY] = 'Nexa';
        // Body font family
        themeVars[PepCustomizationService.FONT_FAMILY_BODY_KEY] = 'Inter';
        // Font sizes
        themeVars[PepCustomizationService.FONT_SIZE_2XS_KEY] =
            '0.625' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FONT_SIZE_XS_KEY] =
            '0.75' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FONT_SIZE_SM_KEY] =
            '0.875' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FONT_SIZE_MD_KEY] =
            '1' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FONT_SIZE_LG_KEY] =
            '1.125' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FONT_SIZE_XL_KEY] =
            '1.25' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FONT_SIZE_2XL_KEY] =
            '1.5' + PepCustomizationService.REM_STRING;
        // Line height
        themeVars[PepCustomizationService.LINE_HEIGHT_2XS_KEY] =
            '0.75' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.LINE_HEIGHT_XS_KEY] =
            '1' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.LINE_HEIGHT_SM_KEY] =
            '1.25' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.LINE_HEIGHT_MD_KEY] =
            '1.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.LINE_HEIGHT_LG_KEY] =
            '1.75' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.LINE_HEIGHT_XL_KEY] =
            '2' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.LINE_HEIGHT_2XL_KEY] =
            '2.25' + PepCustomizationService.REM_STRING;
        // Font weight
        themeVars['--pep-font-weight-normal'] = 400;
        themeVars['--pep-font-weight-bold'] = 600;
        themeVars['--pep-font-weight-bolder'] = 800;
    }
    getDefaultThemeButtonsVariables(themeVars) {
        themeVars['--pep-button-2xs-horizontal-spacing'] =
            '0.25' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-2xs-min-width'] =
            '1' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-2xs-height'] =
            '1' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-2xs-font-size'] =
            'var(' + PepCustomizationService.FONT_SIZE_2XS_KEY + ')';
        // themeVars['--pep-button-2xs-icon-horizontal-spacing'] = '0.125' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-xs-horizontal-spacing'] =
            '0.5' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-xs-min-width'] =
            '1.5' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-xs-height'] =
            '1.5' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-xs-font-size'] =
            'var(' + PepCustomizationService.FONT_SIZE_XS_KEY + ')';
        // themeVars['--pep-button-xs-icon-horizontal-spacing'] = '0.25' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-sm-horizontal-spacing'] =
            '0.75' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-sm-min-width'] =
            '2' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-sm-height'] =
            '2' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-sm-font-size'] =
            'var(' + PepCustomizationService.FONT_SIZE_SM_KEY + ')';
        // themeVars['--pep-button-sm-icon-horizontal-spacing'] = '0.5' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-md-horizontal-spacing'] =
            '1' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-md-min-width'] =
            '2.5' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-md-height'] =
            '2.5' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-md-font-size'] =
            'var(' + PepCustomizationService.FONT_SIZE_MD_KEY + ')';
        // themeVars['--pep-button-md-icon-horizontal-spacing'] = '0.5' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-lg-horizontal-spacing'] =
            '1' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-lg-min-width'] =
            '3' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-lg-height'] =
            '3' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-lg-font-size'] =
            'var(' + PepCustomizationService.FONT_SIZE_LG_KEY + ')';
        // themeVars['--pep-button-lg-icon-horizontal-spacing'] = '0.75' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-xl-horizontal-spacing'] =
            '1.5' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-xl-min-width'] =
            '4' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-xl-height'] =
            '4' + PepCustomizationService.REM_STRING;
        themeVars['--pep-button-xl-font-size'] =
            'var(' + PepCustomizationService.FONT_SIZE_XL_KEY + ')';
        // themeVars['--pep-button-xl-icon-horizontal-spacing'] = '1' + PepCustomizationService.REM_STRING;
    }
    getSystemPrimaryInvertColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_SYSTEM_PRIMARY_INVERT_KEY + '-h'] = '255';
        themeVars[PepCustomizationService.COLOR_SYSTEM_PRIMARY_INVERT_KEY + '-s'] = '100%';
        themeVars[PepCustomizationService.COLOR_SYSTEM_PRIMARY_INVERT_KEY + '-l'] = '100%';
    }
    getSystemPrimaryColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-h'] =
            '0';
        themeVars[PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-s'] =
            '0%';
        themeVars[PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-l'] =
            '10%';
    }
    getSystemSuccessColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_SYSTEM_SUCCESS_KEY + '-h'] =
            '100';
        themeVars[PepCustomizationService.COLOR_SYSTEM_SUCCESS_KEY + '-s'] =
            '100%';
        themeVars[PepCustomizationService.COLOR_SYSTEM_SUCCESS_KEY + '-l'] =
            '25%';
    }
    getSystemCautionColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_SYSTEM_CAUTION_KEY + '-h'] =
            '360';
        themeVars[PepCustomizationService.COLOR_SYSTEM_CAUTION_KEY + '-s'] =
            '100%';
        themeVars[PepCustomizationService.COLOR_SYSTEM_CAUTION_KEY + '-l'] =
            '40%';
    }
    getUserPrimaryColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_USER_PRIMARY_KEY + '-h'] = '78';
        themeVars[PepCustomizationService.COLOR_USER_PRIMARY_KEY + '-s'] =
            '87%';
        themeVars[PepCustomizationService.COLOR_USER_PRIMARY_KEY + '-l'] =
            '27%';
    }
    getUserSecondaryColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_USER_SECONDARY_KEY + '-h'] =
            '77';
        themeVars[PepCustomizationService.COLOR_USER_SECONDARY_KEY + '-s'] =
            '87%';
        themeVars[PepCustomizationService.COLOR_USER_SECONDARY_KEY + '-l'] =
            '42%';
    }
    getWeakColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_WEAK_KEY + '-h'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-h)';
        themeVars[PepCustomizationService.COLOR_WEAK_KEY + '-s'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-s)';
        themeVars[PepCustomizationService.COLOR_WEAK_KEY + '-l'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-l)';
    }
    getRegularColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_REGULAR_KEY + '-h'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-h)';
        themeVars[PepCustomizationService.COLOR_REGULAR_KEY + '-s'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-s)';
        themeVars[PepCustomizationService.COLOR_REGULAR_KEY + '-l'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-l)';
    }
    getStrongColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_STRONG_KEY + '-h'] =
            'var(' + PepCustomizationService.COLOR_USER_PRIMARY_KEY + '-h)';
        themeVars[PepCustomizationService.COLOR_STRONG_KEY + '-s'] =
            'var(' + PepCustomizationService.COLOR_USER_PRIMARY_KEY + '-s)';
        themeVars[PepCustomizationService.COLOR_STRONG_KEY + '-l'] =
            'var(' + PepCustomizationService.COLOR_USER_PRIMARY_KEY + '-l)';
    }
    getTextColorsVariables(themeVars) {
        themeVars[PepCustomizationService.COLOR_TEXT_LINK_KEY + '-h'] = '207';
        themeVars[PepCustomizationService.COLOR_TEXT_LINK_KEY + '-s'] = '76%';
        themeVars[PepCustomizationService.COLOR_TEXT_LINK_KEY + '-l'] = '37%';
    }
    getDefaultThemeColorsVariables(themeVars) {
        // System primary invert colors.
        this.getSystemPrimaryInvertColorsVariables(themeVars);
        // System primary colors.
        this.getSystemPrimaryColorsVariables(themeVars);
        // System success colors.
        this.getSystemSuccessColorsVariables(themeVars);
        // System caution colors.
        this.getSystemCautionColorsVariables(themeVars);
        // User primary colors.
        this.getUserPrimaryColorsVariables(themeVars);
        // User secondary colors.
        this.getUserSecondaryColorsVariables(themeVars);
        // Weak colors.
        this.getWeakColorsVariables(themeVars);
        // Regular colors.
        this.getRegularColorsVariables(themeVars);
        // Strong colors.
        this.getStrongColorsVariables(themeVars);
        // Text colors.
        this.getTextColorsVariables(themeVars);
    }
    getDefaultThemeShadowsVariables(themeVars) {
        const noneOffset = '0';
        themeVars[PepCustomizationService.SHADOW_NONE_OFFSET_KEY] = noneOffset;
        const xsOffset = '0 0.125rem 0.25rem 0';
        themeVars[PepCustomizationService.SHADOW_XS_OFFSET_KEY] = xsOffset;
        const smOffset = '0 0.25rem 0.5rem 0';
        themeVars[PepCustomizationService.SHADOW_SM_OFFSET_KEY] = smOffset;
        const mdOffset = '0 0.5rem 1rem 0';
        themeVars[PepCustomizationService.SHADOW_MD_OFFSET_KEY] = mdOffset;
        const lgOffset = '0 1rem 2rem 0';
        themeVars[PepCustomizationService.SHADOW_LG_OFFSET_KEY] = lgOffset;
        const xlOffset = '0 2rem 4rem 0';
        themeVars[PepCustomizationService.SHADOW_XL_OFFSET_KEY] = xlOffset;
    }
    getDefaultThemeFormCustomizationVariables(themeVars) {
        const formFieldSpacing = 1;
        themeVars[PepCustomizationService.FORM_FIELD_HEIGHT_KEY] =
            '2.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FORM_FIELD_TITLE_HEIGHT_KEY] =
            '1.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FORM_SPACEING_KEY] =
            formFieldSpacing + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.FORM_FIELD_SPACEING_KEY] =
            formFieldSpacing + PepCustomizationService.REM_STRING;
    }
    getDefaultThemeCardCustomizationVariables(themeVars) {
        themeVars[PepCustomizationService.CARD_FIELD_HEIGHT_KEY] =
            '1.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.CARD_SPACEING_KEY] =
            'var(' + PepCustomizationService.SPACING_SIZE_XS_KEY + ')';
        themeVars[PepCustomizationService.CARD_BORDER_RADIUS_KEY] =
            'var(' + PepCustomizationService.BORDER_RADIUS_MD_KEY + ')';
        themeVars[PepCustomizationService.CARD_SHADOW_OFFSET_KEY] =
            'var(' + PepCustomizationService.SHADOW_MD_OFFSET_KEY + ')';
    }
    getDefaultThemeTableCustomizationVariables(themeVars) {
        themeVars[PepCustomizationService.TABLE_FIELD_HEIGHT_KEY] =
            '2' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.TABLE_SPACEING_KEY] =
            '0.5' + PepCustomizationService.REM_STRING;
        themeVars[PepCustomizationService.TABLE_BORDER_RADIUS_KEY] =
            'var(' + PepCustomizationService.BORDER_RADIUS_MD_KEY + ')';
    }
    getDefaultThemeQSCustomizationVariables(themeVars) {
        const qsState = 'regular';
        themeVars[PepCustomizationService.STYLE_QS_KEY] = qsState;
        themeVars[PepCustomizationService.COLOR_QS_KEY + '-h'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-h)';
        themeVars[PepCustomizationService.COLOR_QS_KEY + '-s'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-s)';
        themeVars[PepCustomizationService.COLOR_QS_KEY + '-l'] =
            'var(' + PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY + '-l)';
    }
    getDefaultThemeTopHeaderCustomizationVariables(themeVars) {
        const topHeaderState = 'strong';
        themeVars[PepCustomizationService.STYLE_TOP_HEADER_KEY] = topHeaderState;
        themeVars[PepCustomizationService.COLOR_TOP_HEADER_KEY + '-h'] = '';
        themeVars[PepCustomizationService.COLOR_TOP_HEADER_KEY + '-s'] = '';
        themeVars[PepCustomizationService.COLOR_TOP_HEADER_KEY + '-l'] = '';
    }
    getDefaultThemeVariables() {
        const themeVars = {};
        // Declare default layout.
        this.getDefaultThemeLayoutVariables(themeVars);
        // Declare default border radius.
        this.getDefaultThemeBorderRadiusVariables(themeVars);
        // Declare default spacing.
        this.getDefaultThemeSpacingVariables(themeVars);
        // Declare default fonts size & line height.
        this.getDefaultThemeFontsVariables(themeVars);
        // Declare default buttons.
        this.getDefaultThemeButtonsVariables(themeVars);
        // Declare default colors.
        this.getDefaultThemeColorsVariables(themeVars);
        // Declare default shadows.
        this.getDefaultThemeShadowsVariables(themeVars);
        // Declare default form customization.
        this.getDefaultThemeFormCustomizationVariables(themeVars);
        // Declare default card customization.
        this.getDefaultThemeCardCustomizationVariables(themeVars);
        // Declare default table customization.
        this.getDefaultThemeTableCustomizationVariables(themeVars);
        // Declare default qs customization.
        this.getDefaultThemeQSCustomizationVariables(themeVars);
        // Declare default top header customization.
        this.getDefaultThemeTopHeaderCustomizationVariables(themeVars);
        return themeVars;
    }
    getThemeVariables() {
        const ret = this.getDefaultThemeVariables();
        const themeVars = this.sessionService.getObject(PepCustomizationService.USER_THEME_VARIABLES);
        // Override defaults.
        if (themeVars) {
            for (const key in themeVars) {
                ret[key] = themeVars[key];
            }
        }
        return ret;
    }
    setThemeVariables(themeVariablesToSet = null) {
        const themeVariables = this.getThemeVariables();
        // Override defaults.
        if (themeVariablesToSet) {
            for (const key in themeVariablesToSet) {
                themeVariables[key] = themeVariablesToSet[key];
            }
        }
        this.sessionService.setObject(PepCustomizationService.USER_THEME_VARIABLES, JSON.stringify(themeVariables));
        for (const key in themeVariables) {
            if (this.isFontUrlKey(key)) {
                this.fileService.loadFontStyle(key, themeVariables[key]);
            }
            else {
                document.documentElement.style.setProperty(key, themeVariables[key]);
            }
        }
    }
    isFontUrlKey(key) {
        let res = false;
        if (key === PepCustomizationService.FONT_FAMILY_TITLE_NORMAL_URL_KEY ||
            key === PepCustomizationService.FONT_FAMILY_TITLE_BOLD_URL_KEY ||
            key === PepCustomizationService.FONT_FAMILY_BODY_NORMAL_URL_KEY ||
            key === PepCustomizationService.FONT_FAMILY_BODY_BOLD_URL_KEY) {
            res = true;
        }
        return res;
    }
    getThemeVariable(key) {
        const themeVars = this.getThemeVariables();
        if (themeVars.hasOwnProperty(key)) {
            return themeVars[key];
        }
        else {
            return '';
        }
    }
    getNumberThemeVariable(themeVars, key) {
        if (themeVars && themeVars.hasOwnProperty(key)) {
            // If it's reference to another key.
            if (themeVars[key].indexOf('var(') === 0) {
                key = themeVars[key].substr(4).split(')')[0];
            }
            const tmp = Number(themeVars[key].replace(PepCustomizationService.REM_STRING, ''));
            return tmp;
        }
        else {
            return 0;
        }
    }
    setFooterHeight(height) {
        if (this.footerHeight.getValue() !== height) {
            // it is publishing this value to all the subscribers that have already subscribed to this message
            setTimeout(() => {
                this.footerHeight.next(height);
            }, 0);
        }
        document.documentElement.style.setProperty(PepCustomizationService.FOOTER_HEIGHT_KEY, height + PepCustomizationService.REM_STRING);
    }
    hideFooter() {
        this.setFooterHeight(0);
    }
    showFooter() {
        const themeVars = this.getThemeVariables();
        const res = this.getNumberThemeVariable(themeVars, PepCustomizationService.FOOTER_BAR_SPACING_TOP_KEY) +
            this.getNumberThemeVariable(themeVars, PepCustomizationService.FOOTER_BAR_SPACING_BOTTOM_KEY) +
            this.getNumberThemeVariable(themeVars, PepCustomizationService.TOP_BAR_FIELD_HEIGHT_KEY);
        this.setFooterHeight(res);
    }
    setSettingsWidth(width) {
        if (this.settingsWidth.getValue() !== width) {
            // it is publishing this value to all the subscribers that have already subscribed to this message
            setTimeout(() => {
                this.settingsWidth.next(width);
            }, 0);
        }
        document.documentElement.style.setProperty(PepCustomizationService.SETTINGS_WIDTH_KEY, width + PepCustomizationService.REM_STRING);
    }
    hideSettings() {
        this.setSettingsWidth(0);
    }
    showSettings() {
        const themeVars = this.getThemeVariables();
        const res = this.getNumberThemeVariable(themeVars, PepCustomizationService.SIDE_BAR_WIDTH_KEY);
        this.setSettingsWidth(res);
    }
    getTopBarHeight() {
        const themeVars = this.getThemeVariables();
        const res = this.getNumberThemeVariable(themeVars, PepCustomizationService.TOP_BAR_SPACING_TOP_KEY) +
            this.getNumberThemeVariable(themeVars, PepCustomizationService.TOP_BAR_SPACING_BOTTOM_KEY) +
            this.getNumberThemeVariable(themeVars, PepCustomizationService.TOP_BAR_FIELD_HEIGHT_KEY);
        return res;
    }
    setOldUserTheme(res) {
        // Added user theme.
        const themeObj = {
            Theme: res.Theme || 'default-theme',
        };
        this.sessionService.setObject(PepCustomizationService.USER_THEME, JSON.stringify(themeObj));
        let spinnerColor;
        let brandingFontColor;
        if (this.isLightColor(res.BrandingMainColor)) {
            spinnerColor = PepCustomizationService.DEFAULT_SPINNER_COLOR;
            brandingFontColor = '#222';
        }
        else {
            spinnerColor = res.BrandingMainColor;
            brandingFontColor = PepCustomizationService.DEFAULT_BRANDING_COLOR;
        }
        const color = {
            BrandingSecondaryColor: res.BrandingSecondaryColor,
            BrandingMainColor: res.BrandingMainColor,
            BrandingFontColor: brandingFontColor,
            SpinnerColor: spinnerColor,
        };
        this.sessionService.setObject(PepCustomizationService.USER_COLOR, JSON.stringify(color));
        this.hasCustomHeader =
            res.TopHeaderFiles &&
                res.TopHeaderFiles.length > 0 &&
                res.TopHeaderFiles[0] !== ''
                ? true
                : false;
        this.hasCustomHomepage =
            res.UserRole === 'Buyer' &&
                res.TopHeaderFiles &&
                res.TopHeaderFiles.length > 1 &&
                res.TopHeaderFiles[1] !== '';
    }
    isLegacyColor() {
        const h = document.documentElement.style.getPropertyValue(PepCustomizationService.COLOR_TOP_HEADER_KEY + '-h');
        const s = document.documentElement.style.getPropertyValue(PepCustomizationService.COLOR_TOP_HEADER_KEY + '-s');
        const l = document.documentElement.style.getPropertyValue(PepCustomizationService.COLOR_TOP_HEADER_KEY + '-l');
        return !(h && s && l);
    }
    getLoadingSpinnerColor() {
        let spinnerColor = '';
        if (this.isLegacyColor()) {
            const color = this.sessionService.getObject(PepCustomizationService.USER_COLOR);
            if (color && color.SpinnerColor) {
                spinnerColor = color.SpinnerColor;
            }
            else {
                spinnerColor = PepCustomizationService.DEFAULT_SPINNER_COLOR;
            }
        }
        return spinnerColor;
    }
    getBrandingMainColor() {
        let brandingMainColor = '';
        if (this.isLegacyColor()) {
            const color = this.sessionService.getObject(PepCustomizationService.USER_COLOR);
            if (color && color.BrandingMainColor) {
                brandingMainColor = color.BrandingMainColor;
            }
            else {
                brandingMainColor =
                    PepCustomizationService.DEFAULT_BRANDING_COLOR;
            }
        }
        return brandingMainColor;
    }
    isLightColor(hexColor) {
        if (!hexColor) {
            return;
        }
        let isBright = false;
        let sum = 0;
        const c = hexColor.replace(/^#/, '');
        sum = parseInt(c[0] + c[1], 16);
        sum += parseInt(c[2] + c[3], 16);
        sum += parseInt(c[4] + c[5], 16);
        if (sum > 382.6) {
            // it's bright color
            isBright = true;
        }
        return isBright;
    }
}
PepCustomizationService.STAND_ALONE_FIELD_CLASS_NAME = 'pep-field';
PepCustomizationService.REM_STRING = 'rem';
PepCustomizationService.PX_STRING = 'px';
PepCustomizationService.USER_THEME = 'user_theme';
PepCustomizationService.USER_COLOR = 'user_color';
PepCustomizationService.USER_THEME_VARIABLES = 'user_theme_variables';
// Header height
PepCustomizationService.HEADER_HEIGHT_KEY = '--pep-header-height';
// Side bar variables keys
PepCustomizationService.SIDE_BAR_WIDTH_KEY = '--pep-side-bar-width';
// Settings width
PepCustomizationService.SETTINGS_WIDTH_KEY = '--pep-settings-bar-width';
// Top bar variables keys
PepCustomizationService.TOP_BAR_SPACING_TOP_KEY = '--pep-top-bar-spacing-top';
PepCustomizationService.TOP_BAR_SPACING_BOTTOM_KEY = '--pep-top-bar-spacing-bottom';
PepCustomizationService.TOP_BAR_FIELD_HEIGHT_KEY = '--pep-top-bar-field-height';
// Footer variables keys
PepCustomizationService.FOOTER_HEIGHT_KEY = '--pep-footer-bar-height';
PepCustomizationService.FOOTER_BAR_SPACING_TOP_KEY = '--pep-footer-bar-spacing-top';
PepCustomizationService.FOOTER_BAR_SPACING_BOTTOM_KEY = '--pep-footer-bar-spacing-bottom';
// Main container height
PepCustomizationService.MAIN_HEIGHT_KEY = '--pep-main-height';
// Colors variables keys
PepCustomizationService.COLOR_USER_PRIMARY_KEY = '--pep-color-user-primary';
PepCustomizationService.COLOR_USER_SECONDARY_KEY = '--pep-color-user-secondary';
PepCustomizationService.COLOR_SYSTEM_PRIMARY_INVERT_KEY = '--pep-color-system-primary-invert';
PepCustomizationService.COLOR_SYSTEM_PRIMARY_KEY = '--pep-color-system-primary';
PepCustomizationService.COLOR_TEXT_LINK_KEY = '--pep-color-text-link';
PepCustomizationService.COLOR_SYSTEM_CAUTION_KEY = '--pep-color-system-caution';
PepCustomizationService.COLOR_SYSTEM_SUCCESS_KEY = '--pep-color-system-success';
PepCustomizationService.COLOR_STRONG_KEY = '--pep-color-strong';
PepCustomizationService.COLOR_REGULAR_KEY = '--pep-color-regular';
PepCustomizationService.COLOR_WEAK_KEY = '--pep-color-weak';
// Top header variables keys
PepCustomizationService.COLOR_TOP_HEADER_KEY = '--pep-color-top-header';
PepCustomizationService.STYLE_TOP_HEADER_KEY = '--pep-style-top-header';
// QS variables keys
PepCustomizationService.COLOR_QS_KEY = '--pep-color-qs';
PepCustomizationService.STYLE_QS_KEY = '--pep-style-qs';
// Fonts variables keys
PepCustomizationService.FONT_FAMILY_TITLE_KEY = '--pep-font-family-title';
PepCustomizationService.FONT_FAMILY_TITLE_NORMAL_URL_KEY = '--pep-font-family-title-normal-url';
PepCustomizationService.FONT_FAMILY_TITLE_BOLD_URL_KEY = '--pep-font-family-title-bold-url';
PepCustomizationService.FONT_FAMILY_BODY_KEY = '--pep-font-family-body';
PepCustomizationService.FONT_FAMILY_BODY_NORMAL_URL_KEY = '--pep-font-family-body-normal-url';
PepCustomizationService.FONT_FAMILY_BODY_BOLD_URL_KEY = '--pep-font-family-body-bold-url';
// Border radius variables keys
PepCustomizationService.BORDER_RADIUS_KEY = '--pep-border-radius';
PepCustomizationService.BORDER_RADIUS_SM_KEY = PepCustomizationService.BORDER_RADIUS_KEY + '-sm';
PepCustomizationService.BORDER_RADIUS_MD_KEY = PepCustomizationService.BORDER_RADIUS_KEY + '-md';
PepCustomizationService.BORDER_RADIUS_LG_KEY = PepCustomizationService.BORDER_RADIUS_KEY + '-lg';
// Fonts variables keys
PepCustomizationService.FONT_SIZE_2XS_KEY = '--pep-font-size-2xs';
PepCustomizationService.FONT_SIZE_XS_KEY = '--pep-font-size-xs';
PepCustomizationService.FONT_SIZE_SM_KEY = '--pep-font-size-sm';
PepCustomizationService.FONT_SIZE_MD_KEY = '--pep-font-size-md';
PepCustomizationService.FONT_SIZE_LG_KEY = '--pep-font-size-lg';
PepCustomizationService.FONT_SIZE_XL_KEY = '--pep-font-size-xl';
PepCustomizationService.FONT_SIZE_2XL_KEY = '--pep-font-size-2xl';
// Line height variables keys
PepCustomizationService.LINE_HEIGHT_2XS_KEY = '--pep-line-height-2xs';
PepCustomizationService.LINE_HEIGHT_XS_KEY = '--pep-line-height-xs';
PepCustomizationService.LINE_HEIGHT_SM_KEY = '--pep-line-height-sm';
PepCustomizationService.LINE_HEIGHT_MD_KEY = '--pep-line-height-md';
PepCustomizationService.LINE_HEIGHT_LG_KEY = '--pep-line-height-lg';
PepCustomizationService.LINE_HEIGHT_XL_KEY = '--pep-line-height-xl';
PepCustomizationService.LINE_HEIGHT_2XL_KEY = '--pep-line-height-2xl';
// Shadows offset variables keys
PepCustomizationService.SHADOW_NONE_OFFSET_KEY = '--pep-shadow-none-offset';
PepCustomizationService.SHADOW_XS_OFFSET_KEY = '--pep-shadow-xs-offset';
PepCustomizationService.SHADOW_SM_OFFSET_KEY = '--pep-shadow-sm-offset';
PepCustomizationService.SHADOW_MD_OFFSET_KEY = '--pep-shadow-md-offset';
PepCustomizationService.SHADOW_LG_OFFSET_KEY = '--pep-shadow-lg-offset';
PepCustomizationService.SHADOW_XL_OFFSET_KEY = '--pep-shadow-xl-offset';
// Spacing variables keys
PepCustomizationService.SPACING_SIZE_2XS_KEY = '--pep-spacing-2xs';
PepCustomizationService.SPACING_SIZE_XS_KEY = '--pep-spacing-xs';
PepCustomizationService.SPACING_SIZE_SM_KEY = '--pep-spacing-sm';
PepCustomizationService.SPACING_SIZE_MD_KEY = '--pep-spacing-md';
PepCustomizationService.SPACING_SIZE_LG_KEY = '--pep-spacing-lg';
PepCustomizationService.SPACING_SIZE_XL_KEY = '--pep-spacing-xl';
PepCustomizationService.SPACING_SIZE_2XL_KEY = '--pep-spacing-2xl';
PepCustomizationService.SPACING_SIZE_3XL_KEY = '--pep-spacing-3xl';
PepCustomizationService.SPACING_SIZE_4XL_KEY = '--pep-spacing-4xl';
PepCustomizationService.SPACING_SIZE_5XL_KEY = '--pep-spacing-5xl';
PepCustomizationService.SPACING_SIZE_6XL_KEY = '--pep-spacing-6xl';
PepCustomizationService.SPACING_SIZE_7XL_KEY = '--pep-spacing-7xl';
PepCustomizationService.SPACING_SIZE_8XL_KEY = '--pep-spacing-8xl';
PepCustomizationService.SPACING_SIZE_9XL_KEY = '--pep-spacing-9xl';
// Form variables keys
PepCustomizationService.FORM_FIELD_HEIGHT_KEY = '--pep-form-field-height';
PepCustomizationService.FORM_FIELD_TITLE_HEIGHT_KEY = '--pep-form-field-title-height';
PepCustomizationService.FORM_FIELD_SPACEING_KEY = '--pep-form-field-spacing';
PepCustomizationService.FORM_SPACEING_KEY = '--pep-form-spacing';
// Card variables keys
PepCustomizationService.CARD_FIELD_HEIGHT_KEY = '--pep-card-field-height';
PepCustomizationService.CARD_SPACEING_KEY = '--pep-card-spacing';
PepCustomizationService.CARD_SHADOW_OFFSET_KEY = '--pep-shadow-card-offset';
PepCustomizationService.CARD_BORDER_RADIUS_KEY = '--pep-card-border-radius';
// Table variables keys
PepCustomizationService.TABLE_FIELD_HEIGHT_KEY = '--pep-table-field-height';
PepCustomizationService.TABLE_SPACEING_KEY = '--pep-table-spacing';
PepCustomizationService.TABLE_BORDER_RADIUS_KEY = '--pep-table-border-radius';
PepCustomizationService.DEFAULT_HEADER_HEIGHT = 64; // Default
PepCustomizationService.DEFAULT_SPINNER_COLOR = '#78aa00';
PepCustomizationService.DEFAULT_BRANDING_COLOR = '#fff';
PepCustomizationService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PepCustomizationService_Factory() { return new PepCustomizationService(i0.ɵɵinject(i1.PepSessionService), i0.ɵɵinject(i2.PepFileService), i0.ɵɵinject(i3.FormBuilder)); }, token: PepCustomizationService, providedIn: "root" });
PepCustomizationService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
PepCustomizationService.ctorParameters = () => [
    { type: PepSessionService },
    { type: PepFileService },
    { type: FormBuilder }
];
