// @pepperi-addons/ngx-lib/esm2015/list/list.pipes.js
import { Pipe } from '@angular/core';
export class IsItemSelectedPipe {
    constructor() {
        this.SEPARATOR = ',';
    }
    getUniqItemId(itemId, itemType = '') {
        return itemId + this.SEPARATOR + itemType;
    }
    transform(item, selectionTypeForActions, selectedItemId, isAllSelected, selectedItems, unSelectedItems) {
        let isSelected = false;
        if (selectionTypeForActions === 'single') {
            isSelected = selectedItemId === this.getUniqItemId(item === null || item === void 0 ? void 0 : item.UID, item === null || item === void 0 ? void 0 : item.Type.toString());
        }
        // else if (selectionTypeForActions === 'single-action') {
        //     isSelected = (isAllSelected && !unSelectedItems.has(item?.UID)) || selectedItems.has(item?.UID);
        // }
        else if (selectionTypeForActions === 'multi') {
            if (item === null || item === void 0 ? void 0 : item.IsSelectableForActions) {
                isSelected = (isAllSelected && !unSelectedItems.has(item === null || item === void 0 ? void 0 : item.UID)) || selectedItems.has(item === null || item === void 0 ? void 0 : item.UID);
            }
        }
        return isSelected;
    }
}
IsItemSelectedPipe.decorators = [
    { type: Pipe, args: [{
                name: 'isItemSelected'
            },] }
];
export class IsItemDisabledPipe {
    transform(item, lockItemInnerEvents, printMode) {
        return (printMode || lockItemInnerEvents || (item && !item.IsSelectableForActions));
    }
}
IsItemDisabledPipe.decorators = [
    { type: Pipe, args: [{
                name: 'isItemDisabled'
            },] }
];
