// @pepperi-addons/ngx-lib/esm2015/list/list-views.component.js
import { Component, Injectable, Input, Output, EventEmitter, } from '@angular/core';
export class IPepListView {
}
export class PepListViewsComponent {
    constructor() {
        this._views = null;
        this._currentView = null;
        this.displayType = 'menu';
        this.styleType = 'weak';
        this.sizeType = 'md';
        this.classNames = '';
        this.change = new EventEmitter();
        this.menuItems = null;
        this.currentItem = null;
    }
    set views(value) {
        this._views = value;
        this.menuItems = value.map((opt) => {
            return { key: opt.key, text: opt.title, iconName: opt.iconName };
        });
    }
    get views() {
        return this._views;
    }
    set currentView(value) {
        this._currentView = value;
        if (this.displayType === 'menu') {
            this.currentItem = this.menuItems.find((mi) => mi.key === value.key);
        }
    }
    get currentView() {
        return this._currentView;
    }
    ngOnInit() {
        if (this.displayType === 'buttons') {
            if (this.currentView === null &&
                this.views &&
                this.views.length > 0) {
                this.currentView = this.views[0];
            }
        }
    }
    onMenuItemClicked(menuItemClickEvent) {
        this.currentView = this.views.find((list) => list.key === menuItemClickEvent.source.key);
        this.raiseChangeEvent();
    }
    onViewChanged(view) {
        this.currentView = view;
        this.raiseChangeEvent();
    }
    raiseChangeEvent() {
        this.change.emit({ source: this.currentView });
    }
}
PepListViewsComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-list-views',
                template: "<ng-container *ngIf=\"displayType === 'menu' then menuTemplate else buttonsTemplate\">\n</ng-container>\n<ng-template #menuTemplate>\n    <pep-menu [type]=\"'select'\" [styleType]=\"styleType\" [sizeType]=\"sizeType\" [classNames]=\"classNames\"\n        [items]=\"menuItems\" [selectedItem]=\"currentItem\"\n        [iconName]=\"currentItem?.iconName ? currentItem?.iconName : 'arrow_down'\"\n        (menuItemClick)=\"onMenuItemClicked($event)\"></pep-menu>\n</ng-template>\n<ng-template #buttonsTemplate>\n    <div class=\"buttons-wrapper pep-spacing-element-negative\">\n        <button *ngFor=\"let view of views\" mat-button [title]=\"view.title\" [disabled]=\"view.key === currentView?.key\"\n            class=\"pep-button icon-button pep-spacing-element {{ sizeType }} {{ styleType }} {{ classNames }}\"\n            pepRtlDirection pepMenuBlur (click)=\"onViewChanged(view)\">\n            <mat-icon>\n                <pep-icon name=\"{{ view?.iconName }}\"></pep-icon>\n            </mat-icon>\n        </button>\n    </div>\n</ng-template>",
                styles: [".buttons-wrapper{display:flex}"]
            },] },
    { type: Injectable }
];
PepListViewsComponent.propDecorators = {
    views: [{ type: Input }],
    currentView: [{ type: Input }],
    displayType: [{ type: Input }],
    styleType: [{ type: Input }],
    sizeType: [{ type: Input }],
    classNames: [{ type: Input }],
    change: [{ type: Output }]
};
