// @pepperi-addons/ngx-lib/esm2015/list/list-chooser.component.js
import { Component, Injectable, Input, Output, EventEmitter, } from '@angular/core';
import { PepBreadCrumbItem } from '@pepperi-addons/ngx-lib/bread-crumbs';
export class IPepListChooserOption {
}
export class PepListChooserComponent {
    constructor() {
        this._options = null;
        this._currentList = null;
        this.sizeType = 'md';
        this.change = new EventEmitter();
        this.breadCrumbsItems = null;
        this.menuItems = null;
        this.currentItem = null;
    }
    set options(value) {
        var _a;
        this._options = value;
        this.menuItems = value.map((opt) => {
            return { key: opt.key, text: opt.text };
        });
        if (((_a = this.menuItems) === null || _a === void 0 ? void 0 : _a.length) === 1) {
            this.breadCrumbsItems = [];
            this.breadCrumbsItems.push(new PepBreadCrumbItem({
                key: this.menuItems[0].key,
                text: this.menuItems[0].text,
            }));
        }
    }
    get options() {
        return this._options;
    }
    set currentList(value) {
        this._currentList = value;
        this.currentItem = this.menuItems.find((mi) => mi.key === value.key);
    }
    get currentList() {
        return this._currentList;
    }
    onMenuItemClicked(menuItemClickEvent) {
        this.currentList = this.options.find((list) => list.key === menuItemClickEvent.source.key);
        this.change.emit({ source: this.currentList });
    }
}
PepListChooserComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-list-chooser',
                template: "<ng-container *ngIf=\"menuItems?.length > 1\">\n    <pep-menu [type]=\"'select'\" [sizeType]=\"sizeType\" [items]=\"menuItems\" [selectedItem]=\"currentItem\"\n        [iconName]=\"'arrow_down'\" (menuItemClick)=\"onMenuItemClicked($event)\"></pep-menu>\n</ng-container>\n<ng-container *ngIf=\"menuItems?.length == 1\">\n    <pep-bread-crumbs [items]=\"breadCrumbsItems\" [addSpacing]=\"true\"></pep-bread-crumbs>\n</ng-container>",
                styles: [""]
            },] },
    { type: Injectable }
];
PepListChooserComponent.propDecorators = {
    options: [{ type: Input }],
    currentList: [{ type: Input }],
    sizeType: [{ type: Input }],
    change: [{ type: Output }]
};
