// @pepperi-addons/ngx-lib/esm2015/list/list-carousel.component.js
import { Component, Injectable, Input, Output, EventEmitter, ViewChild, } from '@angular/core';
import { PepLayoutService, PepScreenSizeType, } from '@pepperi-addons/ngx-lib';
import { PepCarouselComponent } from '@pepperi-addons/ngx-lib/carousel';
export class PepListCarouselComponent {
    // deviceHasMouse = false;
    constructor(layoutService) {
        this.layoutService = layoutService;
        this.duration = 1000;
        this.layout = null;
        this.itemsToMove = 3;
        this.lockItemInnerEvents = true;
        this.hideArrowsInSmallScreen = true;
        this._items = null;
        this._itemSize = 'xs';
        this.itemClick = new EventEmitter();
        this.prevDisabled = false;
        this.nextDisabled = false;
        this.PepScreenSizeType = PepScreenSizeType;
        // this.deviceHasMouse = this.layoutService.getDeviceHasMouse();
    }
    set items(value) {
        this._items = value;
        this.moveTo(0);
    }
    get items() {
        return this._items;
    }
    set itemSize(value) {
        this._itemSize = value;
        this.moveTo(0);
    }
    get itemSize() {
        return this._itemSize;
    }
    ngAfterViewInit() {
        this.layoutService.onResize$.subscribe((size) => {
            this.screenSize = size;
        });
        // this.layoutService.onMouseOver$.subscribe((deviceHasMouse: boolean) => {
        //     this.deviceHasMouse = deviceHasMouse;
        // });
    }
    ngOnDestroy() {
        // if (this.itemClick) {
        //     this.itemClick.unsubscribe();
        // }
    }
    itemClicked(item) {
        this.itemClick.emit({ source: item });
    }
    moveLeft() {
        // this.carousel.moveLeft();
        const indexToMove = Math.max(this.carousel.currIndex - this.itemsToMove, 0);
        this.moveTo(indexToMove);
    }
    moveRight() {
        // this.carousel.moveRight();
        const indexToMove = Math.min(this.carousel.currIndex + this.itemsToMove, this.items.length);
        this.moveTo(indexToMove);
    }
    moveTo(index) {
        if (this.carousel) {
            this.carousel.moveTo(index);
        }
    }
    onReachesLeftBound(event) {
        this.prevDisabled = event;
    }
    onReachesRightBound(event) {
        this.nextDisabled = event;
    }
}
PepListCarouselComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-list-carousel',
                template: "<div class=\"carousel-container\">\n    <ng-container *ngIf=\"items?.length > 0; then carouselBlock; else emptyBlock\"></ng-container>\n    <ng-template #emptyBlock>\n        <!-- No items -->\n    </ng-template>\n    <ng-template #carouselBlock>\n        <pep-button *ngIf=\"!(hideArrowsInSmallScreen && screenSize > PepScreenSizeType.SM)\"\n            [iconName]=\"'arrow_left_alt'\" [styleType]=\"'regular'\" [disabled]=\"prevDisabled\" (click)=\"moveLeft()\"\n            class=\"carousel-arrow start-arrow\">\n        </pep-button>\n        <pep-carousel #carousel [xWheelEnabled]=\"false\" [snapDuration]=\"duration\"\n            (reachesLeftBound)=\"onReachesLeftBound($event)\" (reachesRightBound)=\"onReachesRightBound($event)\">\n            <ng-container *ngFor=\"let item of items\">\n                <pep-form pepCarouselItem [layout]=\"layout\" [data]=\"item\" [layoutType]=\"'card'\"\n                    (click)=\"itemClicked(item)\" class=\"card-view \" [lockEvents]=\"lockItemInnerEvents\"\n                    [ngClass]=\"{ 'x-small-card': itemSize === 'xs', 'small-card': itemSize === 'sm'}\">\n                </pep-form>\n            </ng-container>\n        </pep-carousel>\n        <pep-button *ngIf=\"!(hideArrowsInSmallScreen && screenSize > PepScreenSizeType.SM)\"\n            [iconName]=\"'arrow_right_alt'\" [styleType]=\"'regular'\" [disabled]=\"nextDisabled\" (click)=\"moveRight()\"\n            class=\"carousel-arrow end-arrow\">\n        </pep-button>\n    </ng-template>\n</div>",
                styles: [".carousel-container{position:relative;margin:0 calc(.5rem * -1);margin:0 calc(var(--pep-spacing-sm, .5rem) * -1);display:grid}.carousel-container .carousel-arrow{position:absolute;top:50%;transform:translateY(-50%);z-index:1}.carousel-container .carousel-arrow.start-arrow{left:0}.carousel-container .carousel-arrow.end-arrow{right:0}.carousel-container ::ng-deep.carousel-content .card-view{padding:.5rem 0!important;padding:var(--pep-spacing-sm,.5rem) 0!important}.carousel-container ::ng-deep.carousel-content .card-view:first-child{-webkit-padding-start:.5rem!important;padding-inline-start:.5rem!important;-webkit-padding-start:var(--pep-spacing-sm,.5rem)!important;padding-inline-start:var(--pep-spacing-sm,.5rem)!important}.carousel-container ::ng-deep.carousel-content .card-view:last-child{-webkit-padding-end:.5rem!important;padding-inline-end:.5rem!important;-webkit-padding-end:var(--pep-spacing-sm,.5rem)!important;padding-inline-end:var(--pep-spacing-sm,.5rem)!important}"]
            },] },
    { type: Injectable }
];
PepListCarouselComponent.ctorParameters = () => [
    { type: PepLayoutService }
];
PepListCarouselComponent.propDecorators = {
    duration: [{ type: Input }],
    layout: [{ type: Input }],
    itemsToMove: [{ type: Input }],
    lockItemInnerEvents: [{ type: Input }],
    hideArrowsInSmallScreen: [{ type: Input }],
    items: [{ type: Input }],
    itemSize: [{ type: Input }],
    itemClick: [{ type: Output }],
    carousel: [{ type: ViewChild, args: ['carousel', { read: PepCarouselComponent },] }]
};
