// @pepperi-addons/ngx-lib/esm2015/list/list-actions.component.js
import { Component, Injectable, Input, Output, EventEmitter, } from '@angular/core';
export class PepListActionsComponent {
    constructor() {
        this.actions = null;
        this.sizeType = 'md';
        this.xPosition = 'before';
        this.hidden = false;
        this.actionClick = new EventEmitter();
        this.stateChange = new EventEmitter();
        this.menuClick = new EventEmitter();
    }
    onActionClicked(action) {
        this.actionClick.emit(action);
    }
    onStateChanged(menuStateChangeEvent) {
        this.stateChange.emit(menuStateChangeEvent);
    }
}
PepListActionsComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-list-actions',
                template: "<pep-menu [xPosition]=\"xPosition\" [styleType]=\"'strong'\" [sizeType]=\"sizeType\" [items]=\"!hidden ? actions : null\"\n    [hideOnEmptyItems]=\"true\" [iconName]=\"'system_edit'\" (menuItemClick)=\"onActionClicked($event)\"\n    (stateChange)=\"onStateChanged($event)\" (menuClick)=\"menuClick.emit();\">\n</pep-menu>\n",
                styles: [""]
            },] },
    { type: Injectable }
];
PepListActionsComponent.propDecorators = {
    actions: [{ type: Input }],
    sizeType: [{ type: Input }],
    xPosition: [{ type: Input }],
    hidden: [{ type: Input }],
    actionClick: [{ type: Output }],
    stateChange: [{ type: Output }],
    menuClick: [{ type: Output }]
};
