// @pepperi-addons/ngx-lib/esm2015/list/list-sorting.component.js
import { Component, Injectable, Input, Output, EventEmitter, } from '@angular/core';
export class IPepListSortingOption {
}
export class PepListSortingComponent {
    constructor() {
        this._options = null;
        this._currentSorting = null;
        this.sizeType = 'md';
        this.change = new EventEmitter();
        this.menuItems = null;
        this.currentItem = null;
    }
    set options(value) {
        this._options = value;
        this.menuItems = value.map((opt) => {
            return { key: opt.sortBy, text: opt.title, iconName: opt.iconName };
        });
    }
    get options() {
        return this._options;
    }
    set currentSorting(value) {
        this._currentSorting = value;
        this.currentItem = this.menuItems.find((mi) => mi.key === value.sortBy);
    }
    get currentSorting() {
        return this._currentSorting;
    }
    ngOnInit() {
        if (this.currentSorting === null &&
            this.options &&
            this.options.length > 0) {
            this.currentSorting = this.options[0];
        }
    }
    onMenuItemClicked(menuItemClickEvent) {
        this.currentSorting = this.options.find((sorting) => sorting.sortBy === menuItemClickEvent.source.key);
        this.change.emit({ source: this.currentSorting });
    }
}
PepListSortingComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-list-sorting',
                template: "<div class=\"sorting-container\">\n    <div class=\"body-xs sort-by\">\n        <span>{{ 'LIST.SORT_BY' | translate }}</span>\n    </div>\n    <pep-menu [type]=\"'select'\" [sizeType]=\"sizeType\" [items]=\"menuItems\" [selectedItem]=\"currentItem\"\n        [iconName]=\"'arrow_down'\" (menuItemClick)=\"onMenuItemClicked($event)\"></pep-menu>\n</div>",
                styles: [".sorting-container{display:flex;align-items:center}.sorting-container .sort-by{-webkit-margin-end:.5rem;margin-inline-end:.5rem;-webkit-margin-end:var(--pep-spacing-sm,.5rem);margin-inline-end:var(--pep-spacing-sm,.5rem)}"]
            },] },
    { type: Injectable }
];
PepListSortingComponent.propDecorators = {
    options: [{ type: Input }],
    currentSorting: [{ type: Input }],
    sizeType: [{ type: Input }],
    change: [{ type: Output }]
};
