// @pepperi-addons/ngx-lib/esm2015/smart-filters/multi-select-filter/multi-select-filter.component.js
import { ViewChild, } from '@angular/core';
import { Component } from '@angular/core';
import { BaseFilterComponent } from '../common/model/base-filter-component';
import { PepSmartFilterOperators, } from '../common/model/operator';
import { distinctUntilChanged, map, startWith } from 'rxjs/operators';
import { FormControl } from '@angular/forms';
import { VirtualScrollerComponent } from 'ngx-virtual-scroller';
class PepMultiSelectFilterOption {
    constructor() {
        this.selected = false;
    }
}
export class PepMultiSelectFilterComponent extends BaseFilterComponent {
    constructor() {
        super(...arguments);
        this.options = [];
        this.searchControl = new FormControl();
        this.MAX_OPTIONS_TO_SHOW = 6.35;
        this.numberOptionsToShowSearch = 10;
    }
    ngOnInit() {
        var _a;
        super.ngOnInit;
        if (((_a = this.field.options) === null || _a === void 0 ? void 0 : _a.length) > 0) {
            this.options = this.field.options.map((opt) => {
                return { value: opt.value, count: opt.count, selected: false };
            });
        }
        else {
            this.searchControl.disable();
        }
        // Init the selected values from first value.
        const firstControl = this.firstControl;
        if (firstControl.value) {
            this.initOptionsSelectedValues(firstControl.value);
        }
        // Add subscription for the first value change to set the selected options.
        firstControl.valueChanges
            .pipe(this.getDestroyer(), distinctUntilChanged())
            .subscribe((selectedValues) => {
            this.initOptionsSelectedValues(selectedValues);
        });
        // Filter the options by the search control.
        this.filteredOptions$ = this.searchControl.valueChanges.pipe(this.getDestroyer(), startWith(''), map((option) => typeof option === 'string' ? option : option && option.value), map((value) => (value ? this.filterOptions(value) : this.options)));
        // Each time the filter change.
        this.filteredOptions$.subscribe((filterdOptions) => {
            this.calcOptionsHeight(filterdOptions.length);
            setTimeout(() => {
                // 8 is the padding top of the multi-select-options
                this.virtualScroller.scrollToPosition(-8);
            }, 125);
        });
    }
    ngAfterViewInit() {
        // Calc for the first time.
        this.calcOptionsHeight(this.options.length);
    }
    initOptionsSelectedValues(selectedValues) {
        this.options.forEach((opt) => {
            const isValueSelected = selectedValues && selectedValues.includes(opt.value);
            opt.selected = isValueSelected;
        });
    }
    calcOptionsHeight(optionsCount) {
        if (this.optionsContainer) {
            const optionsToShow = optionsCount > this.MAX_OPTIONS_TO_SHOW
                ? this.MAX_OPTIONS_TO_SHOW
                : optionsCount;
            // optionsToShow * 2.25 is 1 option height + 1 is the padding top & bottom of the container.
            const optionsHeight = optionsToShow * 2.25 + 1 + 'rem';
            this.renderer.setStyle(this.optionsContainer.nativeElement, 'height', optionsHeight);
        }
    }
    filterOptions(value) {
        const filterValue = value.toLowerCase();
        return this.options.filter((opt) => opt.value &&
            opt.value.toLowerCase().includes(filterValue.toLowerCase()));
    }
    // Override
    getDefaultOperator() {
        return PepSmartFilterOperators.In;
    }
    // Override
    getFilterValue() {
        const selectedValues = this.options
            .filter((opt) => opt.selected)
            .map((opt) => opt.value);
        const filterValue = {
            first: selectedValues,
        };
        return selectedValues.length > 0 ? filterValue : null;
    }
    // Override
    initFilter() {
        this.options.forEach((opt) => (opt.selected = false));
        this.searchControl.setValue('');
    }
    onOptionChange(option, event) {
        option.selected = event.checked;
        this.firstControl.setValue(this.options.filter((opt) => opt.selected).map((opt) => opt.value), { emitEvent: false });
    }
}
PepMultiSelectFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-multi-select-filter',
                template: "<div [formGroup]=\"form\">\n    <pep-search *ngIf=\"options?.length > numberOptionsToShowSearch\" [searchControl]=\"searchControl\"\n        [triggerOn]=\"'keydown'\" [shrinkInSmallScreen]=\"false\"></pep-search>\n    <div pepRtlClass class=\"multi-select-options pep-border-top pep-border-bottom\"\n        [ngClass]=\"{'with-no-search': options?.length <= numberOptionsToShowSearch}\">\n        <div #optionsContainer class=\"flip-scroll\" pepRtlClass>\n            <virtual-scroller #scroll [items]=\"filteredOptions$ | async\" [parentScroll]=\"optionsContainer\">\n                <mat-checkbox *ngFor=\"let opt of scroll.viewPortItems\"\n                    class=\"pep-checkbox-container pep-input no-default-background md\" [title]=\"opt.value\"\n                    [checked]=\"opt.selected\" (change)=\"onOptionChange(opt, $event)\">\n                    <div class=\"title\">\n                        <span class=\"body-sm ellipsis\">{{ opt.value }}</span>\n                        <span *ngIf=\"opt.count\" class=\"color-dimmed body-xs ellipsis\">({{opt.count}})</span>\n                    </div>\n                </mat-checkbox>\n                <ng-container *ngIf=\"scroll.viewPortItems?.length === 0\">\n                    <span class=\"no-options body-sm ellipsis\" [title]=\"field.name\">\n                        {{ ( options.length > 0 ? \"SMART_FILTERS.NO_SEARCH_RESULTS\" : \"SMART_FILTERS.NO_OPTIONS\") |\n                        translate }}\n                    </span>\n                </ng-container>\n            </virtual-scroller>\n        </div>\n    </div>\n</div>",
                styles: [".multi-select-options{margin:.5rem calc(.25rem * -1) 0;margin:var(--pep-spacing-sm,.5rem) calc(var(--pep-spacing-xs, .25rem) * -1) 0}.multi-select-options.with-no-search{margin:calc(.5rem * -1) calc(.25rem * -1) 0;margin:calc(var(--pep-spacing-sm, .5rem) * -1) calc(var(--pep-spacing-xs, .25rem) * -1) 0}.multi-select-options .no-options{display:flex;padding:.5rem .75rem;padding:var(--pep-spacing-sm,.5rem) var(--pep-spacing-md,.75rem)}.multi-select-options .flip-scroll{padding:.5rem .25rem;padding:var(--pep-spacing-sm,.5rem) var(--pep-spacing-xs,.25rem);min-height:3.25rem;max-height:34.75rem;direction:rtl}.multi-select-options .flip-scroll.rtl,.multi-select-options .flip-scroll virtual-scroller{direction:ltr}.multi-select-options .flip-scroll.rtl virtual-scroller{direction:rtl}.multi-select-options .pep-checkbox-container{height:2.25rem;-webkit-padding-end:0;padding-inline-end:0}.multi-select-options .pep-checkbox-container ::ng-deep.mat-checkbox-layout{width:100%;height:100%;grid-template-columns:auto 1fr}.multi-select-options .pep-checkbox-container ::ng-deep.mat-checkbox-layout .mat-checkbox-label{align-self:center}.multi-select-options .pep-checkbox-container .title{display:flex;justify-content:space-between;align-items:center;margin:0 .25rem;margin:0 var(--pep-spacing-xs,.25rem)}"]
            },] }
];
PepMultiSelectFilterComponent.propDecorators = {
    optionsContainer: [{ type: ViewChild, args: ['optionsContainer',] }],
    virtualScroller: [{ type: ViewChild, args: [VirtualScrollerComponent,] }]
};
