// @pepperi-addons/ngx-lib/esm2015/color/color.component.js
import { Component, Input, Output, Renderer2, ElementRef, EventEmitter, } from '@angular/core';
import { PepCustomizationService, DEFAULT_HORIZONTAL_ALIGNMENT, } from '@pepperi-addons/ngx-lib';
import { PepDialogService } from '@pepperi-addons/ngx-lib/dialog';
import { PepColorPickerComponent } from './color-picker.component';
export class PepColorComponent {
    constructor(dialogService, renderer, element) {
        this.dialogService = dialogService;
        this.renderer = renderer;
        this.element = element;
        this.key = '';
        this.value = '';
        this.label = '';
        this.disabled = false;
        this.xAlignment = DEFAULT_HORIZONTAL_ALIGNMENT;
        this.rowSpan = 1;
        this.type = 'any';
        this.showTitle = true;
        this.showAAComplient = true;
        this.layoutType = 'form';
        this.valueChange = new EventEmitter();
    }
    ngOnInit() {
        this.renderer.addClass(this.element.nativeElement, PepCustomizationService.STAND_ALONE_FIELD_CLASS_NAME);
    }
    ngOnDestroy() {
        //
    }
    changeColor(value) {
        this.value = value;
        this.valueChange.emit(value);
    }
    chooseColor() {
        const dialogRef = this.dialogService.openDialog(PepColorPickerComponent, {
            value: this.value,
            type: this.type,
            showAAComplient: this.showAAComplient,
        });
        dialogRef.afterClosed().subscribe((value) => {
            if (value !== undefined && value !== null) {
                this.changeColor(value);
            }
        });
    }
}
PepColorComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-color',
                template: "<pep-field-title [label]=\"label\" [disabled]=\"disabled\" [xAlignment]=\"xAlignment\" [showTitle]=\"showTitle\">\n</pep-field-title>\n<div [id]=\"key\" class=\"pep-color-container\" [ngClass]=\"{ 'right-alignment': xAlignment == 'right' }\">\n    <div pepRtlDirection (click)=\"!disabled ? chooseColor() : ''\" class=\"pep-color pep-input\"\n        [ngClass]=\"{disable: disabled, 'one-row': rowSpan === 1}\"\n        style=\"background: {{ value ? value + '!important' : 'transparent' }}; opacity: {{ disabled && value ? 0.12 : 1}};\">\n        <button *ngIf=\"!disabled\" class=\"pep-button icon-button xs weak invert mat-button btn-edit\" tabindex=\"-1\"\n            pepButtonBlur mat-button>\n            <mat-icon>\n                <pep-icon name=\"system_edit\"></pep-icon>\n            </mat-icon>\n        </button>\n    </div>\n</div>",
                styles: [":host{display:grid}.pep-color-container .pep-color,.pep-color-container .pepperi-color,.pepperi-color-container .pep-color,.pepperi-color-container .pepperi-color{padding:0;display:grid}.pep-color-container .pep-color .btn-edit,.pep-color-container .pepperi-color .btn-edit,.pepperi-color-container .pep-color .btn-edit,.pepperi-color-container .pepperi-color .btn-edit{width:calc(2.5rem - 1rem);width:calc(var(--pep-form-field-height, 2.5rem) - var(--pep-spacing-lg, 1rem));min-width:unset;padding:0;margin:0 .75rem;margin:0 var(--pep-spacing-md,.75rem);justify-self:end;align-self:center}.pep-color-container.right-alignment .pep-color .btn-edit,.pep-color-container.right-alignment .pepperi-color .btn-edit,.pepperi-color-container.right-alignment .pep-color .btn-edit,.pepperi-color-container.right-alignment .pepperi-color .btn-edit{justify-self:start}"]
            },] }
];
PepColorComponent.ctorParameters = () => [
    { type: PepDialogService },
    { type: Renderer2 },
    { type: ElementRef }
];
PepColorComponent.propDecorators = {
    key: [{ type: Input }],
    value: [{ type: Input }],
    label: [{ type: Input }],
    disabled: [{ type: Input }],
    xAlignment: [{ type: Input }],
    rowSpan: [{ type: Input }],
    type: [{ type: Input }],
    showTitle: [{ type: Input }],
    showAAComplient: [{ type: Input }],
    layoutType: [{ type: Input }],
    valueChange: [{ type: Output }]
};
