// @pepperi-addons/ngx-lib/esm2015/textbox/textbox-validation.directive.js
import { Directive, ElementRef, HostListener, Input, } from '@angular/core';
import { PepValidatorService, } from '@pepperi-addons/ngx-lib';
export class PepTextboxValidationDirective {
    constructor(hostElement, validatorService) {
        this.hostElement = hostElement;
        this.validatorService = validatorService;
        this.type = 'text';
        this.previousValue = '';
    }
    ngOnInit() {
        //
    }
    isNumber() {
        return this.isInteger() || this.isDecimal();
    }
    isInteger() {
        return this.type === 'int';
    }
    isDecimal() {
        return (this.type === 'currency' ||
            this.type === 'percentage' ||
            this.type === 'real');
    }
    isPhone() {
        return this.type === 'phone';
    }
    isText() {
        return (this.type === 'email' ||
            this.type === 'link' ||
            this.type === 'text');
    }
    onChange(e) {
        this.validateValue(this.hostElement.nativeElement.value);
    }
    onKeyDown(e) {
        const originalValue = e.target['value'];
        if (this.isNumber()) {
            // save value before keydown event
            this.previousValue = originalValue;
            const isNumber = this.validatorService.isNumber(e, this.isDecimal());
            if (isNumber)
                return;
            else
                e.preventDefault();
        }
        else if (this.isPhone()) {
            const isPhone = this.validatorService.isPhone(e);
            if (isPhone)
                return;
            else
                e.preventDefault();
        }
    }
    validateValue(value) {
        if (this.isNumber()) {
            const newValue = this.validatorService.validateNumber(value, this.isDecimal());
            this.hostElement.nativeElement['value'] = newValue ? newValue : 0;
        }
        else if (this.isPhone()) {
            // test phone with regular expression, when
            // phone is invalid, replace it with the previousValue
            const valid = this.validatorService.validatePhone(value);
            this.hostElement.nativeElement['value'] = valid
                ? value
                : this.previousValue;
        }
        else if (this.isText()) {
            this.hostElement.nativeElement['value'] = value;
        }
    }
}
PepTextboxValidationDirective.decorators = [
    { type: Directive, args: [{
                selector: '[pepTextboxValidation]',
            },] }
];
PepTextboxValidationDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: PepValidatorService }
];
PepTextboxValidationDirective.propDecorators = {
    type: [{ type: Input }],
    onChange: [{ type: HostListener, args: ['change', ['$event'],] }],
    onKeyDown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
};
