// @pepperi-addons/ngx-lib/esm2015/smart-filters/boolean-filter/boolean-filter.component.js
import { Component } from '@angular/core';
import { BaseFilterComponent } from '../common/model/base-filter-component';
import { PepSmartFilterOperators, } from '../common/model/operator';
export class PepBooleanFilterComponent extends BaseFilterComponent {
    // Override
    getDefaultOperator() {
        return PepSmartFilterOperators.Equals;
    }
    // Override
    getFilterValue() {
        const filterValue = { first: this.firstControl.value || false };
        return filterValue;
    }
}
PepBooleanFilterComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-boolean-filter',
                template: "<div [formGroup]=\"form\">\n    <mat-radio-group class=\"radio-group\" [formControlName]=\"firstControlKey\">\n        <mat-radio-button class=\"pep-radio-button md pep-input no-default-background\" value=\"true\">\n            {{ 'CHECKBOX.TRUE' | translate }}\n        </mat-radio-button>\n        <mat-radio-button class=\"pep-radio-button md pep-input no-default-background\" value=\"false\">\n            {{ 'CHECKBOX.FALSE' | translate }}\n        </mat-radio-button>\n    </mat-radio-group>\n</div>",
                styles: [".radio-group{display:flex;flex-direction:column}.radio-group .pep-radio-button.pep-input{height:2.25rem;display:flex}"]
            },] }
];
