// @pepperi-addons/ngx-lib/esm2015/date/date.model.js
import { Inject, Injectable, Optional } from "@angular/core";
import { MAT_DATE_LOCALE } from "@angular/material/core";
import { MomentDateAdapter } from "@angular/material-moment-adapter";
import { MomentDatetimeAdapter } from '@mat-datetimepicker/moment';
export class MomentUtcDateAdapter extends MomentDateAdapter {
    constructor(dateLocale) {
        super(dateLocale);
    }
}
MomentUtcDateAdapter.decorators = [
    { type: Injectable }
];
MomentUtcDateAdapter.ctorParameters = () => [
    { type: String, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_LOCALE,] }] }
];
export class MomentUtcDateTimeAdapter extends MomentDatetimeAdapter {
    constructor(dateLocale) {
        super(dateLocale, { strict: false, useUtc: false }, new MomentUtcDateAdapter(dateLocale));
    }
}
MomentUtcDateTimeAdapter.decorators = [
    { type: Injectable }
];
MomentUtcDateTimeAdapter.ctorParameters = () => [
    { type: String, decorators: [{ type: Optional }, { type: Inject, args: [MAT_DATE_LOCALE,] }] }
];
export const MY_DATE_FORMATS = {
    parse: {
        dateInput: 'L',
        monthInput: 'MMMM',
        timeInput: 'LT',
        datetimeInput: 'L LT',
    },
    display: {
        dateInput: 'L',
        monthInput: 'MMMM',
        timeInput: 'LT',
        datetimeInput: 'L LT',
        monthYearLabel: 'MMM YYYY',
        dateA11yLabel: 'LL',
        monthYearA11yLabel: 'MMMM YYYY',
        popupHeaderDateLabel: 'ddd, DD MMM',
    },
};
