// @pepperi-addons/ngx-lib/esm2015/quantity-selector/quantity-selector-validation.directive.js
import { Directive, ElementRef, HostListener, Input, } from '@angular/core';
import { PepValidatorService } from '@pepperi-addons/ngx-lib';
export class PepQuantitySelectorValidationDirective {
    constructor(hostElement, validatorService) {
        this.hostElement = hostElement;
        this.validatorService = validatorService;
        this.previousValue = '';
    }
    ngOnInit() {
        //
    }
    onChange(e) {
        this.validateValue(this.hostElement.nativeElement.value);
    }
    onKeyDown(e) {
        const originalValue = e.target['value'];
        // save value before keydown event
        this.previousValue = originalValue;
        const isNumber = this.validatorService.isNumber(e, this.allowDecimal);
        if (isNumber)
            return;
        else
            e.preventDefault();
    }
    validateValue(value) {
        const newValue = this.validatorService.validateNumber(value, this.allowDecimal);
        this.hostElement.nativeElement['value'] = newValue ? newValue : 0;
    }
}
PepQuantitySelectorValidationDirective.decorators = [
    { type: Directive, args: [{
                selector: '[pepQsValidation]',
            },] }
];
PepQuantitySelectorValidationDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: PepValidatorService }
];
PepQuantitySelectorValidationDirective.propDecorators = {
    allowDecimal: [{ type: Input }],
    onChange: [{ type: HostListener, args: ['change', ['$event'],] }],
    onKeyDown: [{ type: HostListener, args: ['keydown', ['$event'],] }]
};
