// @pepperi-addons/ngx-lib/bundles/pepperi-addons-ngx-lib-separator.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@pepperi-addons/ngx-lib')) :
    typeof define === 'function' && define.amd ? define('@pepperi-addons/ngx-lib/separator', ['exports', '@angular/core', '@angular/common', '@pepperi-addons/ngx-lib'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['pepperi-addons'] = global['pepperi-addons'] || {}, global['pepperi-addons']['ngx-lib'] = global['pepperi-addons']['ngx-lib'] || {}, global['pepperi-addons']['ngx-lib'].separator = {}), global.ng.core, global.ng.common, global['pepperi-addons']['ngx-lib']));
}(this, (function (exports, core, common, ngxLib) { 'use strict';

    var PepSeparatorComponent = /** @class */ (function () {
        function PepSeparatorComponent(renderer, element) {
            this.renderer = renderer;
            this.element = element;
            this.key = '';
            this.label = '';
            this.xAlignment = ngxLib.DEFAULT_HORIZONTAL_ALIGNMENT;
            this.controlType = 'separator';
            this.form = null;
            this.layoutType = 'form';
            this._visible = true;
            this.standAlone = false;
        }
        Object.defineProperty(PepSeparatorComponent.prototype, "visible", {
            get: function () {
                return this._visible;
            },
            set: function (visible) {
                this._visible = visible;
                if (visible) {
                    this.renderer.removeClass(this.element.nativeElement, 'hidden-element');
                }
                else {
                    this.renderer.addClass(this.element.nativeElement, 'hidden-element');
                }
            },
            enumerable: false,
            configurable: true
        });
        PepSeparatorComponent.prototype.ngOnInit = function () {
            if (this.form === null) {
                this.standAlone = true;
                this.renderer.addClass(this.element.nativeElement, ngxLib.PepCustomizationService.STAND_ALONE_FIELD_CLASS_NAME);
            }
        };
        return PepSeparatorComponent;
    }());
    PepSeparatorComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'pep-separator',
                    template: "<ng-container *ngIf=\"layoutType === 'form'\">\n    <div class=\"pep-separator-field wrap pep-border-bottom align-{{ xAlignment }}\" title=\"{{ label }}\">\n        <span class=\"block-with-text body-md\">{{ label }}</span>\n    </div>\n</ng-container>\n\n<ng-container *ngIf=\"layoutType === 'card'\">\n    <div class=\"pep-separator-field pep-small-separator-field pep-border-bottom text-align-{{ xAlignment }}\"\n        dir=\"{{ xAlignment == 'right' ? 'rtl' : 'ltr' }}\" title=\"{{ label }}\">\n        <span class=\"small-block-with-text body-sm\">{{ label }}</span>\n    </div>\n</ng-container>\n\n<ng-container *ngIf=\"layoutType === 'table'\">\n    <label class=\"cl\" title=\"{{ label }}\">{{ label }}</label>\n</ng-container>",
                    changeDetection: core.ChangeDetectionStrategy.OnPush,
                    styles: [":host{height:inherit;display:grid}:host>*{align-self:center}"]
                },] }
    ];
    PepSeparatorComponent.ctorParameters = function () { return [
        { type: core.Renderer2 },
        { type: core.ElementRef }
    ]; };
    PepSeparatorComponent.propDecorators = {
        key: [{ type: core.Input }],
        label: [{ type: core.Input }],
        xAlignment: [{ type: core.Input }],
        form: [{ type: core.Input }],
        layoutType: [{ type: core.Input }],
        visible: [{ type: core.Input }]
    };

    var PepSeparatorModule = /** @class */ (function () {
        function PepSeparatorModule() {
        }
        return PepSeparatorModule;
    }());
    PepSeparatorModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [common.CommonModule],
                    exports: [PepSeparatorComponent],
                    declarations: [PepSeparatorComponent],
                },] }
    ];

    /*
     * Public API Surface of ngx-lib/separator
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.PepSeparatorComponent = PepSeparatorComponent;
    exports.PepSeparatorModule = PepSeparatorModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
