// @pepperi-addons/ngx-lib/bundles/pepperi-addons-ngx-lib-page-layout.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/material/core'), require('@pepperi-addons/ngx-lib')) :
    typeof define === 'function' && define.amd ? define('@pepperi-addons/ngx-lib/page-layout', ['exports', '@angular/core', '@angular/common', '@angular/material/core', '@pepperi-addons/ngx-lib'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['pepperi-addons'] = global['pepperi-addons'] || {}, global['pepperi-addons']['ngx-lib'] = global['pepperi-addons']['ngx-lib'] || {}, global['pepperi-addons']['ngx-lib']['page-layout'] = {}), global.ng.core, global.ng.common, global.ng.material.core, global['pepperi-addons']['ngx-lib']));
}(this, (function (exports, core, common, core$1, ngxLib) { 'use strict';

    var PepPageLayoutComponent = /** @class */ (function () {
        function PepPageLayoutComponent(customizationService, layoutService, cdRef, hostElement, renderer) {
            this.customizationService = customizationService;
            this.layoutService = layoutService;
            this.cdRef = cdRef;
            this.hostElement = hostElement;
            this.renderer = renderer;
            this.addPadding = true;
            this.showShadow = false;
            this.renderer.addClass(this.hostElement.nativeElement, 'pep-page-layout');
        }
        PepPageLayoutComponent.prototype.ngAfterViewInit = function () {
            var _this = this;
            this.layoutService.onResize$.subscribe(function (size) {
                _this.screenSize = size;
            });
        };
        PepPageLayoutComponent.prototype.ngAfterContentInit = function () {
            //
        };
        return PepPageLayoutComponent;
    }());
    PepPageLayoutComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'pep-page-layout',
                    template: "<div class=\"pep-page-header-layout\">\n    <ng-content select=\"[pep-header-area]\"></ng-content>\n</div>\n<div class=\"pep-page-side-layout\" [ngClass]=\"{ 'show-shadow': showShadow }\">\n    <ng-content select=\"pep-side-bar\"></ng-content>\n    <ng-content select=\"[pep-side-area]\"></ng-content>\n</div>\n<div class=\"pep-page-main-layout\" [ngClass]=\"{ 'add-padding': addPadding }\">\n    <div class=\"pep-top-area\">\n        <ng-content select=\"[pep-top-area]\"></ng-content>\n    </div>\n    <div class=\"pep-main-area\">\n        <ng-content select=\"[pep-main-area]\"></ng-content>\n    </div>\n</div>",
                    styles: [":host.pep-page-layout{display:grid;grid-template-columns:auto 1fr;grid-template-rows:auto 1fr;grid-template-areas:\"header-layout  header-layout\" \"side-layout    main-layout\";height:100%;width:100%;overflow:hidden}:host.pep-page-layout .pep-page-header-layout{grid-area:header-layout;z-index:10}:host.pep-page-layout .pep-page-side-layout{grid-area:side-layout;overflow:hidden;height:100%;z-index:5}:host.pep-page-layout .pep-page-side-layout.show-shadow{-webkit-padding-end:1rem;padding-inline-end:1rem;-webkit-padding-end:var(--pep-spacing-lg,1rem);padding-inline-end:var(--pep-spacing-lg,1rem)}:host.pep-page-layout .pep-page-main-layout{grid-area:main-layout;overflow:hidden;height:100%;z-index:1;padding-bottom:4.5rem;padding-bottom:var(--pep-footer-bar-height,4.5rem);display:grid;grid-template-rows:auto 1fr;grid-template-areas:\"top-area\" \"main-area\"}:host.pep-page-layout .pep-page-main-layout.add-padding{padding-inline:1rem;padding-inline:var(--pep-spacing-lg,1rem)}:host.pep-page-layout .pep-page-main-layout .pep-top-area{grid-area:top-area}:host.pep-page-layout .pep-page-main-layout .pep-main-area{grid-area:main-area;overflow:auto;height:100%;display:block}:host.pep-page-layout .pep-page-main-layout .pep-main-area>*{height:inherit}"]
                },] }
    ];
    PepPageLayoutComponent.ctorParameters = function () { return [
        { type: ngxLib.PepCustomizationService },
        { type: ngxLib.PepLayoutService },
        { type: core.ChangeDetectorRef },
        { type: core.ElementRef },
        { type: core.Renderer2 }
    ]; };
    PepPageLayoutComponent.propDecorators = {
        addPadding: [{ type: core.Input }],
        showShadow: [{ type: core.Input }]
    };

    var PepPageLayoutModule = /** @class */ (function () {
        function PepPageLayoutModule() {
        }
        return PepPageLayoutModule;
    }());
    PepPageLayoutModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule,
                        // Material modules,
                        core$1.MatCommonModule,
                        // ngx-lib modules
                        ngxLib.PepNgxLibModule,
                    ],
                    exports: [PepPageLayoutComponent],
                    declarations: [PepPageLayoutComponent],
                },] }
    ];

    /*
     * Public API Surface of ngx-lib/page-layout
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.PepPageLayoutComponent = PepPageLayoutComponent;
    exports.PepPageLayoutModule = PepPageLayoutModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
