// @pepperi-addons/ngx-lib/bundles/pepperi-addons-ngx-lib-menu.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('@angular/material/core'), require('@angular/material/form-field'), require('@angular/material/menu'), require('@angular/material/icon'), require('@pepperi-addons/ngx-lib'), require('@pepperi-addons/ngx-lib/icon'), require('@angular/animations')) :
    typeof define === 'function' && define.amd ? define('@pepperi-addons/ngx-lib/menu', ['exports', '@angular/core', '@angular/common', '@angular/material/core', '@angular/material/form-field', '@angular/material/menu', '@angular/material/icon', '@pepperi-addons/ngx-lib', '@pepperi-addons/ngx-lib/icon', '@angular/animations'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global['pepperi-addons'] = global['pepperi-addons'] || {}, global['pepperi-addons']['ngx-lib'] = global['pepperi-addons']['ngx-lib'] || {}, global['pepperi-addons']['ngx-lib'].menu = {}), global.ng.core, global.ng.common, global.ng.material.core, global.ng.material.formField, global.ng.material.menu, global.ng.material.icon, global['pepperi-addons']['ngx-lib'], global['pepperi-addons']['ngx-lib'].icon, global.ng.animations));
}(this, (function (exports, core, common, core$1, formField, menu, icon$1, ngxLib, icon, animations) { 'use strict';

    var PepMenuItemParent = /** @class */ (function () {
        function PepMenuItemParent(data) {
            this.parent = null;
            this.key = data.key;
            this.parent = data.parent;
        }
        return PepMenuItemParent;
    }());
    var PepMenuItem = /** @class */ (function () {
        function PepMenuItem(data) {
            this.disabled = false;
            this.hidden = false;
            this.iconName = null;
            this.type = 'regular';
            this.children = null;
            this.parent = null;
            Object.assign(this, data);
        }
        return PepMenuItem;
    }());

    var PepMenuComponent = /** @class */ (function () {
        function PepMenuComponent(layoutService) {
            var _this = this;
            this.layoutService = layoutService;
            this.text = null;
            /**
             * The icon name to show on the menu. look in (@link icon -> All icons)
             *
             * @type {PepIconType} See {@link PepIconType}
             * @memberof PepMenuComponent
             */
            this.iconName = icon.pepIconSystemMenu.name;
            this.type = 'action';
            this.styleType = 'weak';
            this.sizeType = 'md';
            this.classNames = '';
            this.xPosition = 'after';
            this.hideOnEmptyItems = false;
            this._items = null;
            this.selectedItem = null;
            this.disabled = false;
            this.stateChange = new core.EventEmitter();
            this.menuItemClick = new core.EventEmitter();
            this.menuClick = new core.EventEmitter();
            this.state = 'hidden';
            this.PepScreenSizeType = ngxLib.PepScreenSizeType;
            this.displayText = null;
            this.layoutService.onResize$.subscribe(function (size) {
                _this.screenSize = size;
            });
        }
        Object.defineProperty(PepMenuComponent.prototype, "items", {
            get: function () {
                return this._items;
            },
            set: function (items) {
                this.setItemsParent(items);
                this._items = items;
            },
            enumerable: false,
            configurable: true
        });
        PepMenuComponent.prototype.setChildrenParent = function (item, parent) {
            var _this = this;
            item.parent = parent;
            if (item.children && item.children.length > 0) {
                item.children.forEach(function (child) {
                    _this.setChildrenParent(child, new PepMenuItemParent(item));
                });
            }
        };
        PepMenuComponent.prototype.setItemsParent = function (items) {
            var _this = this;
            if (items) {
                items.forEach(function (item) {
                    item.parent = null;
                    if (item.children && item.children.length > 0) {
                        item.children.forEach(function (child) {
                            _this.setChildrenParent(child, new PepMenuItemParent(item));
                        });
                    }
                });
            }
        };
        PepMenuComponent.prototype.updateText = function () {
            if (this.type === 'select' || this.type === 'action-select') {
                this.displayText = this.selectedItem
                    ? this.selectedItem.text
                    : this.text;
            }
            else {
                this.displayText = this.text;
            }
        };
        PepMenuComponent.prototype.ngOnChanges = function (changes) {
            if (this.hideOnEmptyItems) {
                this.state =
                    !this.disabled &&
                        this.items &&
                        this.items.filter(function (item) { return !item.disabled; }).length > 0
                        ? 'visible'
                        : 'hidden';
            }
            else {
                this.state = 'visible';
            }
            if (this.type === 'select') {
                if (this.selectedItem === null &&
                    this.items &&
                    this.items.length > 0) {
                    this.selectedItem = this.items[0];
                }
            }
            this.updateText();
        };
        PepMenuComponent.prototype.ngOnDestroy = function () {
            // if (this.menuItemClick) {
            //     this.menuItemClick.unsubscribe();
            // }
        };
        PepMenuComponent.prototype.onMenuClicked = function (event) {
            this.menuClick.emit();
        };
        // private manipulateData(menuItem: PepMenuItem) {
        //     menuItem.children = null;
        //     if (menuItem.parent) {
        //         menuItem.parent = new PepMenuItem(menuItem.parent);
        //         this.manipulateData(menuItem.parent);
        //     }
        // }
        PepMenuComponent.prototype.onMenuItemClicked = function (click) {
            this.selectedItem = click.source;
            this.updateText();
            // Manipulate click data because the data is dupplicate in parent.children
            // const tmp = new PepMenuItem(click.source);
            // this.manipulateData(tmp);
            // click.source = tmp;
            this.menuItemClick.emit(click);
        };
        PepMenuComponent.prototype.animationDone = function () {
            var _this = this;
            if (this.state === 'hidden') {
                setTimeout(function () {
                    _this.stateChange.emit({ state: _this.state });
                }, 500);
            }
        };
        PepMenuComponent.prototype.animationStart = function () {
            if (this.state === 'visible') {
                this.stateChange.emit({ state: this.state });
            }
        };
        return PepMenuComponent;
    }());
    PepMenuComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'pep-menu',
                    template: "<div class=\"menu-container\" dir=\"{{ layoutService.isRtl() ? 'rtl' : 'ltr' }}\">\n\n    <ng-container *ngIf=\"hideOnEmptyItems then animationTemplate else regularTemplate\">\n    </ng-container>\n    <ng-template #animationTemplate>\n        <button mat-button class=\"pep-button {{ sizeType }} {{ styleType }} {{ classNames }}\"\n            [ngClass]=\"{ 'icon-button': !displayText || screenSize > PepScreenSizeType.SM, 'pep-button-limited-width': screenSize < PepScreenSizeType.XS  }\"\n            pepRtlDirection pepMenuBlur (click)=\"onMenuClicked($event)\" [@slideInOut]=\"state\"\n            (@slideInOut.done)=\"animationDone()\" (@slideInOut.start)=\"animationStart()\"\n            [matMenuTriggerFor]=\"menu.childMenu\">\n            <ng-container *ngTemplateOutlet=\"triggerTemplate\">\n            </ng-container>\n        </button>\n        <pep-menu-item #menu [type]=\"type\" [items]=\"items\" [selectedItem]=\"selectedItem\" [xPosition]=\"xPosition\"\n            (menuItemClick)=\"onMenuItemClicked($event)\">\n        </pep-menu-item>\n    </ng-template>\n    <ng-template #regularTemplate>\n        <ng-container *ngIf=\"items then menuTemplate else buttonTemplate\"></ng-container>\n        <ng-template #menuTemplate>\n            <button mat-button class=\"pep-button {{ sizeType }} {{ styleType }} {{ classNames }}\"\n                [ngClass]=\"{ 'icon-button': !displayText || screenSize > PepScreenSizeType.SM, 'pep-button-limited-width': screenSize < PepScreenSizeType.XS, disabled: disabled }\"\n                pepRtlDirection pepMenuBlur (menuClick)=\"onMenuClicked($event)\" [disabled]=\"disabled\"\n                [matMenuTriggerFor]=\"menu.childMenu\">\n                <ng-container *ngTemplateOutlet=\"triggerTemplate\">\n                </ng-container>\n            </button>\n            <pep-menu-item #menu [type]=\"type\" [items]=\"items\" [selectedItem]=\"selectedItem\" [xPosition]=\"xPosition\"\n                (menuItemClick)=\"onMenuItemClicked($event)\"></pep-menu-item>\n        </ng-template>\n        <ng-template #buttonTemplate>\n            <button mat-button class=\"pep-button {{ sizeType }} {{ styleType }} {{ classNames }}\"\n                [ngClass]=\"{ 'icon-button': !displayText || screenSize > PepScreenSizeType.SM, 'pep-button-limited-width': screenSize < PepScreenSizeType.XS, disabled: disabled }\"\n                pepRtlDirection pepMenuBlur (menuClick)=\"onMenuClicked($event)\" [disabled]=\"disabled\">\n                <ng-container *ngTemplateOutlet=\"triggerTemplate\">\n                </ng-container>\n            </button>\n        </ng-template>\n    </ng-template>\n</div>\n\n<ng-template #triggerTemplate>\n    <span *ngIf=\"displayText && screenSize <= PepScreenSizeType.SM\" class=\"ellipsis\"\n        [ngClass]=\"{ 'button-title-with-icon': iconName }\" [title]=\"displayText\">\n        {{ displayText }}\n    </span>\n    <mat-icon *ngIf=\"iconName\">\n        <pep-icon name=\"{{ iconName }}\"></pep-icon>\n    </mat-icon>\n</ng-template>",
                    animations: [
                        animations.trigger('slideInOut', [
                            animations.state('hidden', animations.style({
                                width: '0px',
                                padding: '0',
                                margin: '0',
                                minWidth: '0px',
                                opacity: 0,
                            })),
                            animations.state('visible', animations.style({
                                width: 'inherit',
                                opacity: 1,
                            })),
                            animations.transition('close => open', animations.animate('500ms ease-in')),
                            animations.transition('open => close', animations.animate('500ms ease-out')),
                        ]),
                    ],
                    styles: [".menu-container .pep-button{display:flex;align-items:center;justify-content:center}"]
                },] }
    ];
    PepMenuComponent.ctorParameters = function () { return [
        { type: ngxLib.PepLayoutService }
    ]; };
    PepMenuComponent.propDecorators = {
        text: [{ type: core.Input }],
        iconName: [{ type: core.Input }],
        type: [{ type: core.Input }],
        styleType: [{ type: core.Input }],
        sizeType: [{ type: core.Input }],
        classNames: [{ type: core.Input }],
        xPosition: [{ type: core.Input }],
        hideOnEmptyItems: [{ type: core.Input }],
        items: [{ type: core.Input }],
        selectedItem: [{ type: core.Input }],
        disabled: [{ type: core.Input }],
        stateChange: [{ type: core.Output }],
        menuItemClick: [{ type: core.Output }],
        menuClick: [{ type: core.Output }]
    };

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (Object.prototype.hasOwnProperty.call(b, p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __rest(s, e) {
        var t = {};
        for (var p in s)
            if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
                t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }
    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
            r = Reflect.decorate(decorators, target, key, desc);
        else
            for (var i = decorators.length - 1; i >= 0; i--)
                if (d = decorators[i])
                    r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }
    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); };
    }
    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function")
            return Reflect.metadata(metadataKey, metadataValue);
    }
    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }
    var __createBinding = Object.create ? (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        Object.defineProperty(o, k2, { enumerable: true, get: function () { return m[k]; } });
    }) : (function (o, m, k, k2) {
        if (k2 === undefined)
            k2 = k;
        o[k2] = m[k];
    });
    function __exportStar(m, o) {
        for (var p in m)
            if (p !== "default" && !Object.prototype.hasOwnProperty.call(o, p))
                __createBinding(o, m, p);
    }
    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m)
            return m.call(o);
        if (o && typeof o.length === "number")
            return {
                next: function () {
                    if (o && i >= o.length)
                        o = void 0;
                    return { value: o && o[i++], done: !o };
                }
            };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    /** @deprecated */
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }
    /** @deprecated */
    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++)
            s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    }
    function __spreadArray(to, from) {
        for (var i = 0, il = from.length, j = to.length; i < il; i++, j++)
            to[j] = from[i];
        return to;
    }
    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }
    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n])
            i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try {
            step(g[n](v));
        }
        catch (e) {
            settle(q[0][3], e);
        } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length)
            resume(q[0][0], q[0][1]); }
    }
    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }
    function __asyncValues(o) {
        if (!Symbol.asyncIterator)
            throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function (v) { resolve({ value: v, done: d }); }, reject); }
    }
    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) {
            Object.defineProperty(cooked, "raw", { value: raw });
        }
        else {
            cooked.raw = raw;
        }
        return cooked;
    }
    ;
    var __setModuleDefault = Object.create ? (function (o, v) {
        Object.defineProperty(o, "default", { enumerable: true, value: v });
    }) : function (o, v) {
        o["default"] = v;
    };
    function __importStar(mod) {
        if (mod && mod.__esModule)
            return mod;
        var result = {};
        if (mod != null)
            for (var k in mod)
                if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k))
                    __createBinding(result, mod, k);
        __setModuleDefault(result, mod);
        return result;
    }
    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }
    function __classPrivateFieldGet(receiver, state, kind, f) {
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a getter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot read private member from an object whose class did not declare it");
        return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
    }
    function __classPrivateFieldSet(receiver, state, value, kind, f) {
        if (kind === "m")
            throw new TypeError("Private method is not writable");
        if (kind === "a" && !f)
            throw new TypeError("Private accessor was defined without a setter");
        if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
            throw new TypeError("Cannot write private member to an object whose class did not declare it");
        return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
    }

    var PepInternalMenuItem = /** @class */ (function (_super) {
        __extends(PepInternalMenuItem, _super);
        function PepInternalMenuItem() {
            var _this = _super.apply(this, __spreadArray([], __read(arguments))) || this;
            _this.selected = false;
            return _this;
        }
        return PepInternalMenuItem;
    }(PepMenuItem));
    var PepMenuItemComponent = /** @class */ (function () {
        function PepMenuItemComponent() {
            this.type = 'action';
            this.parent = null;
            this.items = [];
            this._selectedItem = null;
            this.xPosition = 'after';
            this.subMenuIconName = icon.pepIconArrowRight.name;
            this.menuItemClick = new core.EventEmitter();
        }
        Object.defineProperty(PepMenuItemComponent.prototype, "selectedItem", {
            get: function () {
                return this._selectedItem;
            },
            set: function (selectedItem) {
                this.updateSelectedItem(selectedItem);
            },
            enumerable: false,
            configurable: true
        });
        PepMenuItemComponent.prototype.ngOnDestroy = function () {
            // if (this.menuItemClick) {
            //     this.menuItemClick.unsubscribe();
            // }
        };
        PepMenuItemComponent.prototype.clearSelectedItem = function (selectedItem) {
            var _this = this;
            if (this.items) {
                this.items.forEach(function (item) {
                    item.selected = (selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.key) === item.key;
                    if (item.children) {
                        item.children.forEach(function (child) {
                            _this.clearSelectedChildren(child, selectedItem);
                        });
                    }
                });
            }
        };
        PepMenuItemComponent.prototype.clearSelectedChildren = function (item, selectedItem) {
            var _this = this;
            item.selected = (selectedItem === null || selectedItem === void 0 ? void 0 : selectedItem.key) === item.key;
            if (item.children) {
                item.children.forEach(function (child) {
                    _this.clearSelectedChildren(child, selectedItem);
                });
            }
        };
        PepMenuItemComponent.prototype.selectParentAndChildren = function (items, parentsKeys) {
            if (parentsKeys.length > 0) {
                var key_1 = parentsKeys.pop();
                var selectedParent = items.find(function (item) { return item.key === key_1; });
                if (selectedParent) {
                    selectedParent.selected = true;
                    this.selectParentAndChildren(selectedParent.children, parentsKeys);
                }
            }
        };
        PepMenuItemComponent.prototype.updateSelectedParent = function (parentItem, parentsKeys) {
            parentsKeys.push(parentItem.key);
            if (parentItem.parent) {
                this.updateSelectedParent(parentItem.parent, parentsKeys);
            }
            else {
                this.selectParentAndChildren(this.items, parentsKeys);
            }
        };
        PepMenuItemComponent.prototype.updateSelectedItem = function (selectedItem) {
            this.clearSelectedItem(selectedItem);
            this._selectedItem = selectedItem;
            if (selectedItem && selectedItem.parent) {
                this.updateSelectedParent(selectedItem.parent, []);
            }
        };
        PepMenuItemComponent.prototype.onChildClicked = function (click) {
            // click.source.selected = true;
            this.menuItemClick.emit(click);
        };
        PepMenuItemComponent.prototype.onMenuItemClicked = function (item) {
            // item.selected = true;
            this.menuItemClick.emit({ source: item });
        };
        return PepMenuItemComponent;
    }());
    PepMenuItemComponent.decorators = [
        { type: core.Component, args: [{
                    selector: 'pep-menu-item',
                    template: "<mat-menu #childMenu=\"matMenu\" xPosition=\"{{ xPosition }}\" class=\"pep-{{type}}-menu\">\n    <ng-container *ngFor=\"let menuItem of items\">\n        <ng-container *ngIf=\"!menuItem.hidden\">\n            <ng-container *ngIf=\"menuItem.type === 'splitter'; then splitterBlock; else buttonBlock\"></ng-container>\n            <ng-template #splitterBlock>\n                <div mat-menu-item class=\"splitter\"></div>\n            </ng-template>\n\n            <ng-template #buttonBlock>\n                <ng-container\n                    *ngIf=\"type !== 'select' && menuItem.children?.length > 0; then childrenBlock; else regularBlock\">\n                </ng-container>\n                <ng-template #childrenBlock>\n                    <button mat-menu-item (click)=\"onMenuItemClicked(menuItem)\" [matMenuTriggerFor]=\"menu.childMenu\"\n                        title=\"{{ menuItem.text }}\" pepRtlDirection [disabled]=\"menuItem.disabled\"\n                        [ngClass]=\"{ selected: type === 'action-select' && menuItem.selected, disabled: menuItem.disabled }\">\n                        <ng-container *ngTemplateOutlet=\"itemTemplate; context: {\n                            itemIconName: menuItem.iconName,\n                            itemText: menuItem.text,\n                            itemType: menuItem.type,\n                            hasChildren: true\n                        }\"></ng-container>\n                    </button>\n                    <pep-menu-item #menu [type]=\"type\" [items]=\"menuItem.children\" [parent]=\"menuItem\"\n                        (menuItemClick)=\"onChildClicked($event)\">\n                    </pep-menu-item>\n                </ng-template>\n                <ng-template #regularBlock>\n                    <button mat-menu-item (click)=\"onMenuItemClicked(menuItem)\" title=\"{{ menuItem.text }}\"\n                        pepRtlDirection [disabled]=\"menuItem.disabled\"\n                        [ngClass]=\"{ selected: (type === 'select' || type === 'action-select') && menuItem.selected, disabled: menuItem.disabled }\">\n                        <ng-container *ngTemplateOutlet=\"itemTemplate; context: {\n                            itemIconName: menuItem.iconName,\n                            itemText: menuItem.text,\n                            itemType: menuItem.type,\n                            hasChildren: false\n                        }\"></ng-container>\n                    </button>\n                </ng-template>\n            </ng-template>\n        </ng-container>\n    </ng-container>\n</mat-menu>\n\n<ng-template #itemTemplate let-itemIconName=\"itemIconName\" let-itemText=\"itemText\" let-itemType=\"itemType\"\n    let-hasChildren=\"hasChildren\">\n    <ng-container>\n        <mat-icon *ngIf=\"itemIconName\" class=\"pull-left flip\">\n            <pep-icon name=\"{{itemIconName}}\"></pep-icon>\n        </mat-icon>\n        <span *ngIf=\"itemText\" class=\"pull-left flip title ellipsis\"\n            [ngClass]=\"{ 'spacing-from-icon': itemIconName, 'spacing-from-submenu-trigger': hasChildren }\"\n            [title]=\"itemText\">\n            {{ itemText }}\n        </span>\n        <mat-icon *ngIf=\"hasChildren\" class=\"pull-left flip submenu-trigger\"\n            [ngClass]=\"{ 'rotate': xPosition === 'before' }\">\n            <pep-icon name=\"{{ subMenuIconName }}\"></pep-icon>\n        </mat-icon>\n    </ng-container>\n</ng-template>",
                    styles: [".mat-menu-item-submenu-trigger:after{display:none}.mat-menu-item .title{margin:0 .25rem;margin:0 var(--pep-spacing-xs,.25rem)}.mat-menu-item .title.spacing-from-icon{margin:0 .5rem;margin:0 var(--pep-spacing-sm,.5rem)}.mat-menu-item .title.spacing-from-submenu-trigger{-webkit-padding-end:1rem;padding-inline-end:1rem;-webkit-padding-end:var(--pep-spacing-lg,1rem);padding-inline-end:var(--pep-spacing-lg,1rem)}.mat-menu-item .mat-icon{margin:0}.mat-menu-item .submenu-trigger{position:absolute;right:.5rem;right:var(--pep-spacing-sm,.5rem);width:1rem}.mat-menu-item .submenu-trigger.rotate{transform:rotate(180deg)}.mat-menu-item.splitter{height:1px!important;margin:.5rem 0;margin:var(--pep-spacing-sm,.5rem) 0}.mat-menu-item[dir=rtl] .submenu-trigger{right:unset;left:.5rem;left:var(--pep-spacing-sm,.5rem);transform:rotate(180deg)}.mat-menu-item[dir=rtl] .submenu-trigger.rotate{transform:rotate(0deg)}.mat-menu-item-submenu-trigger:after{width:unset;height:unset;transform:unset}"]
                },] }
    ];
    PepMenuItemComponent.propDecorators = {
        type: [{ type: core.Input }],
        parent: [{ type: core.Input }],
        items: [{ type: core.Input }],
        selectedItem: [{ type: core.Input }],
        xPosition: [{ type: core.Input }],
        subMenuIconName: [{ type: core.Input }],
        childMenu: [{ type: core.ViewChild, args: ['childMenu', { static: true },] }],
        menuItemClick: [{ type: core.Output }]
    };

    var PepMenuModule = /** @class */ (function () {
        function PepMenuModule(pepIconRegistry) {
            this.pepIconRegistry = pepIconRegistry;
            this.pepIconRegistry.registerIcons([
                icon.pepIconSystemMenu,
                icon.pepIconArrowRight,
            ]);
        }
        return PepMenuModule;
    }());
    PepMenuModule.decorators = [
        { type: core.NgModule, args: [{
                    imports: [
                        common.CommonModule,
                        // Material modules,
                        core$1.MatCommonModule,
                        formField.MatFormFieldModule,
                        menu.MatMenuModule,
                        icon$1.MatIconModule,
                        // ngx-lib modules
                        ngxLib.PepNgxLibModule,
                        icon.PepIconModule,
                    ],
                    exports: [PepMenuComponent],
                    declarations: [PepMenuComponent, PepMenuItemComponent],
                },] }
    ];
    PepMenuModule.ctorParameters = function () { return [
        { type: icon.PepIconRegistry }
    ]; };

    /*
     * Public API Surface of ngx-lib/menu
     */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.PepMenuComponent = PepMenuComponent;
    exports.PepMenuItem = PepMenuItem;
    exports.PepMenuItemParent = PepMenuItemParent;
    exports.PepMenuModule = PepMenuModule;
    exports.ɵa = PepMenuItemComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
