// @pepperi-addons/ngx-lib/esm2015/core/portal/portal.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class PepPortalService {
    constructor() {
        this.targets = new Map();
    }
    addTarget(targetName, viewContainer) {
        this.targets.set(targetName, viewContainer);
    }
    attach(targetName, template) {
        const target = this.getTarget(targetName);
        if (target) {
            target.createEmbeddedView(template);
        }
    }
    clear(targetName) {
        const target = this.getTarget(targetName);
        if (target) {
            target.clear();
        }
    }
    getTarget(targetName) {
        return this.targets.has(targetName)
            ? this.targets.get(targetName)
            : null;
    }
}
PepPortalService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PepPortalService_Factory() { return new PepPortalService(); }, token: PepPortalService, providedIn: "root" });
PepPortalService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
PepPortalService.ctorParameters = () => [];
