// @pepperi-addons/ngx-lib/esm2015/dialog/dialog.service.js
import { Injectable } from '@angular/core';
import { MatDialog, MatDialogConfig, } from '@angular/material/dialog';
import { PepLayoutService } from '@pepperi-addons/ngx-lib';
import { Overlay } from '@angular/cdk/overlay';
import { PepDefaultDialogComponent } from './default-dialog.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/material/dialog";
import * as i2 from "@pepperi-addons/ngx-lib";
import * as i3 from "@angular/cdk/overlay";
export class PepDialogService {
    constructor(dialog, layoutService, overlay) {
        this.dialog = dialog;
        this.layoutService = layoutService;
        this.overlay = overlay;
    }
    fixConfigIfNeeded(data, config) {
        if (config.disableClose) {
            if (((!data.showClose || !data.showHeader) && !data.showFooter) ||
                (data.actionsType === 'custom' &&
                    (data.actionButtons === null ||
                        data.actionButtons.length === 0))) {
                config.disableClose = false;
            }
        }
    }
    getDialogConfig(options = {}, size = 'regular') {
        var _a;
        const dialogConfig = new MatDialogConfig();
        dialogConfig.autoFocus = false;
        dialogConfig.direction = this.layoutService.isRtl() ? 'rtl' : 'ltr';
        dialogConfig.disableClose = (_a = options.disableClose) !== null && _a !== void 0 ? _a : true;
        dialogConfig.panelClass = ['pep-dialog', options.panelClass || ''];
        if (size === 'inline') {
            dialogConfig.height = options.height || 'auto';
            dialogConfig.maxWidth = options.maxWidth || '100vw';
            dialogConfig.maxHeight = options.maxHeight || '100vh';
            dialogConfig.minWidth = options.minWidth || '300px';
        }
        dialogConfig.panelClass.push(size);
        dialogConfig.scrollStrategy = this.overlay.scrollStrategies.noop();
        // dialogConfig.scrollStrategy = this.overlay.scrollStrategies.block();
        return dialogConfig;
    }
    openDefaultDialog(data, config = null) {
        if (!config) {
            config = this.getDialogConfig();
        }
        this.fixConfigIfNeeded(data, config);
        config.data = data;
        const dialogRef = this.dialog.open(PepDefaultDialogComponent, config);
        return dialogRef;
    }
    openDialog(componentOrTemplateRef, data = {}, config = null) {
        if (!config) {
            config = this.getDialogConfig();
        }
        this.fixConfigIfNeeded(data, config);
        config.data = data;
        const dialogRef = this.dialog.open(componentOrTemplateRef, config);
        return dialogRef;
    }
}
PepDialogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PepDialogService_Factory() { return new PepDialogService(i0.ɵɵinject(i1.MatDialog), i0.ɵɵinject(i2.PepLayoutService), i0.ɵɵinject(i3.Overlay)); }, token: PepDialogService, providedIn: "root" });
PepDialogService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
PepDialogService.ctorParameters = () => [
    { type: MatDialog },
    { type: PepLayoutService },
    { type: Overlay }
];
