// @pepperi-addons/ngx-lib/esm2015/dialog/dialog.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { MatCommonModule } from '@angular/material/core';
import { MatButtonModule } from '@angular/material/button';
import { MatIconModule } from '@angular/material/icon';
import { MatDialogModule } from '@angular/material/dialog';
import { PepNgxLibModule } from '@pepperi-addons/ngx-lib';
import { PepIconModule, PepIconRegistry, pepIconSystemClose, } from '@pepperi-addons/ngx-lib/icon';
import { PepDefaultDialogComponent } from './default-dialog.component';
import { PepDialogComponent } from './dialog.component';
export class PepDialogModule {
    constructor(pepIconRegistry) {
        this.pepIconRegistry = pepIconRegistry;
        this.pepIconRegistry.registerIcons([pepIconSystemClose]);
    }
}
PepDialogModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    // Material modules,
                    MatCommonModule,
                    MatButtonModule,
                    MatIconModule,
                    MatDialogModule,
                    // ngx-lib modules
                    PepNgxLibModule,
                    PepIconModule,
                ],
                exports: [PepDialogComponent],
                declarations: [PepDefaultDialogComponent, PepDialogComponent],
                entryComponents: [PepDefaultDialogComponent],
            },] }
];
PepDialogModule.ctorParameters = () => [
    { type: PepIconRegistry }
];
