// @pepperi-addons/ngx-lib/esm2015/dialog/dialog.component.js
import { Component, Input } from '@angular/core';
export class PepDialogComponent {
    constructor() {
        this.showClose = true;
        this.showHeader = true;
        this.showFooter = true;
    }
}
PepDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-dialog',
                template: "<div mat-dialog-title *ngIf=\"showHeader\" class=\"dialog-header-container\">\n    <span class=\"dialog-title\" *ngIf=\"title?.length\">\n        {{ title }}\n    </span>\n    <button mat-button *ngIf=\"showClose\" [mat-dialog-close]=\"null\" class=\"pep-button icon-button md weak dialog-close\">\n        <mat-icon>\n            <pep-icon name=\"system_close\">\n            </pep-icon>\n        </mat-icon>\n    </button>\n</div>\n<div mat-dialog-content>\n    <ng-content select=\"[pep-dialog-content]\"></ng-content>\n</div>\n<div mat-dialog-actions *ngIf=\"showFooter\" class=\"pep-border-top\">\n    <ng-content select=\"[pep-dialog-actions]\"></ng-content>\n</div>",
                styles: [""]
            },] }
];
PepDialogComponent.propDecorators = {
    title: [{ type: Input }],
    showClose: [{ type: Input }],
    showHeader: [{ type: Input }],
    showFooter: [{ type: Input }]
};
