// @pepperi-addons/ngx-lib/esm2015/core/http/services/loader.service.js
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/operators';
import * as i0 from "@angular/core";
export class PepLoaderService {
    constructor() {
        this.counter = 0;
        this.showLoaderSubject = new BehaviorSubject(false);
        // Raise custom event for showing the loader.
        this.onChanged$.subscribe((show) => {
            this.showLoaderChanged(show);
        });
    }
    get onChanged$() {
        return this.showLoaderSubject
            .asObservable()
            .pipe(distinctUntilChanged());
    }
    showLoaderChanged(show) {
        const eventData = {
            detail: {
                showLoader: show,
            },
        };
        const event = new CustomEvent('toggleLoader', eventData);
        window.dispatchEvent(event);
    }
    show() {
        this.counter++;
        this.showLoaderSubject.next(true);
    }
    hide() {
        this.counter--;
        if (this.counter < 0) {
            this.counter = 0;
        }
        if (this.counter === 0) {
            this.showLoaderSubject.next(false);
        }
    }
}
PepLoaderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PepLoaderService_Factory() { return new PepLoaderService(); }, token: PepLoaderService, providedIn: "root" });
PepLoaderService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
PepLoaderService.ctorParameters = () => [];
