// @pepperi-addons/ngx-lib/esm2015/smart-filters/common/model/creator.js
import { PepSmartFilterBooleanField, PepSmartFilterDateField, PepSmartFilterMultiSelectField, PepSmartFilterIntField, PepSmartFilterDateTimeField, PepSmartFilterPercentageField, PepSmartFilterCurrencyField, PepSmartFilterRealField, } from './field';
function createBooleanSmartFilterField(data) {
    return new PepSmartFilterBooleanField(data);
}
function createDateSmartFilterField(data) {
    return new PepSmartFilterDateField(data);
}
function createDateTimeSmartFilterField(data) {
    return new PepSmartFilterDateTimeField(data);
}
function createMultiSelectSmartFilterField(data) {
    return new PepSmartFilterMultiSelectField(data);
}
function createIntSmartFilterField(data) {
    return new PepSmartFilterIntField(data);
}
function createRealSmartFilterField(data) {
    return new PepSmartFilterRealField(data);
}
function createCurrencySmartFilterField(data) {
    return new PepSmartFilterCurrencyField(data);
}
function createPercentageSmartFilterField(data) {
    return new PepSmartFilterPercentageField(data);
}
export function createSmartFilterField(data, type) {
    let field;
    switch (type) {
        case 'boolean':
            field = createBooleanSmartFilterField(data);
            break;
        case 'date':
            field = createDateSmartFilterField(data);
            break;
        case 'date-time':
            field = createDateTimeSmartFilterField(data);
            break;
        case 'multi-select':
            field = createMultiSelectSmartFilterField(data);
            break;
        case 'int':
            field = createIntSmartFilterField(data);
            break;
        case 'real':
            field = createRealSmartFilterField(data);
            break;
        case 'currency':
            field = createCurrencySmartFilterField(data);
            break;
        case 'percentage':
            field = createPercentageSmartFilterField(data);
            break;
        default:
            field = createMultiSelectSmartFilterField(data);
            break;
    }
    return field;
}
export function createSmartFilter(fieldId, operator, first, second, operatorUnit) {
    return {
        fieldId: fieldId,
        operator: operator,
        operatorUnit: operatorUnit,
        value: {
            first: first,
            second: second,
        },
    };
}
