// @pepperi-addons/ngx-lib/esm2015/page-layout/page-layout.component.js
import { ChangeDetectorRef, ElementRef, Renderer2, } from '@angular/core';
import { Component, Input, } from '@angular/core';
import { PepCustomizationService, PepLayoutService, } from '@pepperi-addons/ngx-lib';
export class PepPageLayoutComponent {
    constructor(customizationService, layoutService, cdRef, hostElement, renderer) {
        this.customizationService = customizationService;
        this.layoutService = layoutService;
        this.cdRef = cdRef;
        this.hostElement = hostElement;
        this.renderer = renderer;
        this.addPadding = true;
        this.showShadow = false;
        this.renderer.addClass(this.hostElement.nativeElement, 'pep-page-layout');
    }
    ngAfterViewInit() {
        this.layoutService.onResize$.subscribe((size) => {
            this.screenSize = size;
        });
    }
    ngAfterContentInit() {
        //
    }
}
PepPageLayoutComponent.decorators = [
    { type: Component, args: [{
                selector: 'pep-page-layout',
                template: "<div class=\"pep-page-header-layout\">\n    <ng-content select=\"[pep-header-area]\"></ng-content>\n</div>\n<div class=\"pep-page-side-layout\" [ngClass]=\"{ 'show-shadow': showShadow }\">\n    <ng-content select=\"pep-side-bar\"></ng-content>\n    <ng-content select=\"[pep-side-area]\"></ng-content>\n</div>\n<div class=\"pep-page-main-layout\" [ngClass]=\"{ 'add-padding': addPadding }\">\n    <div class=\"pep-top-area\">\n        <ng-content select=\"[pep-top-area]\"></ng-content>\n    </div>\n    <div class=\"pep-main-area\">\n        <ng-content select=\"[pep-main-area]\"></ng-content>\n    </div>\n</div>",
                styles: [":host.pep-page-layout{display:grid;grid-template-columns:auto 1fr;grid-template-rows:auto 1fr;grid-template-areas:\"header-layout  header-layout\" \"side-layout    main-layout\";height:100%;width:100%;overflow:hidden}:host.pep-page-layout .pep-page-header-layout{grid-area:header-layout;z-index:10}:host.pep-page-layout .pep-page-side-layout{grid-area:side-layout;overflow:hidden;height:100%;z-index:5}:host.pep-page-layout .pep-page-side-layout.show-shadow{-webkit-padding-end:1rem;padding-inline-end:1rem;-webkit-padding-end:var(--pep-spacing-lg,1rem);padding-inline-end:var(--pep-spacing-lg,1rem)}:host.pep-page-layout .pep-page-main-layout{grid-area:main-layout;overflow:hidden;height:100%;z-index:1;padding-bottom:4.5rem;padding-bottom:var(--pep-footer-bar-height,4.5rem);display:grid;grid-template-rows:auto 1fr;grid-template-areas:\"top-area\" \"main-area\"}:host.pep-page-layout .pep-page-main-layout.add-padding{padding-inline:1rem;padding-inline:var(--pep-spacing-lg,1rem)}:host.pep-page-layout .pep-page-main-layout .pep-top-area{grid-area:top-area}:host.pep-page-layout .pep-page-main-layout .pep-main-area{grid-area:main-area;overflow:auto;height:100%;display:block}:host.pep-page-layout .pep-page-main-layout .pep-main-area>*{height:inherit}"]
            },] }
];
PepPageLayoutComponent.ctorParameters = () => [
    { type: PepCustomizationService },
    { type: PepLayoutService },
    { type: ChangeDetectorRef },
    { type: ElementRef },
    { type: Renderer2 }
];
PepPageLayoutComponent.propDecorators = {
    addPadding: [{ type: Input }],
    showShadow: [{ type: Input }]
};
